// @hotstuff-labs/ts-sdk/dist/apis/exchange.api.js
import { privateKeyToAccount } from "viem/accounts";
import { signAction, NonceManager } from "../utils";
import { EXCHANGE_OP_CODES, } from "../methods";
export class ExchangeClient {
    constructor(args) {
        this.transport = args.transport;
        this.wallet = args.wallet;
        this.nonce = args.nonce ?? new NonceManager().getNonce;
    }
    /**---Account Management Methods---*/
    async addAgent(params, execute = true, opts) {
        let nonce = await this.nonce();
        const agentAccount = privateKeyToAccount(params.agentPrivateKey);
        const agentSignature = await signAction({
            wallet: agentAccount,
            action: {
                signer: params.signer,
                nonce: nonce,
            },
            txType: EXCHANGE_OP_CODES["addAgent"],
        }, {
            isTestnet: this.transport.isTestnet,
        });
        params = {
            agentName: params.agentName,
            agent: params.agent,
            forAccount: params.forAccount,
            signature: agentSignature,
            validUntil: params.validUntil,
            nonce: nonce,
        };
        return this._executeAction({ action: "addAgent", params }, opts?.signal, execute);
    }
    async revokeAgent(params, opts) {
        return this._executeAction({ action: "revokeAgent", params }, opts?.signal);
    }
    async updatePerpInstrumentLeverage(params, opts) {
        return this._executeAction({ action: "updatePerpInstrumentLeverage", params }, opts?.signal);
    }
    async approveBrokerFee(params, opts) {
        return this._executeAction({ action: "approveBrokerFee", params }, opts?.signal);
    }
    async createReferralCode(params, opts) {
        return this._executeAction({ action: "createReferralCode", params }, opts?.signal);
    }
    async setReferrer(params, opts) {
        return this._executeAction({ action: "setReferrer", params }, opts?.signal);
    }
    async claimReferralRewards(params, opts) {
        return this._executeAction({ action: "claimReferralRewards", params }, opts?.signal);
    }
    /**---Trading Methods---*/
    async placeOrder(params, opts) {
        return this._executeAction({ action: "placeOrder", params }, opts?.signal);
    }
    async cancelByOid(params, opts) {
        return this._executeAction({ action: "cancelByOid", params }, opts?.signal);
    }
    async cancelByCloid(params, opts) {
        return this._executeAction({ action: "cancelByCloid", params }, opts?.signal);
    }
    async cancelByInstrument(params, opts) {
        return this._executeAction({ action: "cancelByInstrument", params }, opts?.signal);
    }
    async cancelAll(params, opts) {
        return this._executeAction({ action: "cancelAll", params }, opts?.signal);
    }
    /**---Collateral Transfer Methods---*/
    async accountSpotWithdrawRequest(params, opts) {
        return this._executeAction({ action: "accountSpotWithdrawRequest", params }, opts?.signal);
    }
    async accountDerivativeWithdrawRequest(params, opts) {
        return this._executeAction({ action: "accountDerivativeWithdrawRequest", params }, opts?.signal);
    }
    async accountSpotBalanceTransferRequest(params, opts) {
        return this._executeAction({ action: "accountSpotBalanceTransferRequest", params }, opts?.signal);
    }
    async accountDerivativeBalanceTransferRequest(params, opts) {
        return this._executeAction({ action: "accountDerivativeBalanceTransferRequest", params }, opts?.signal);
    }
    async accountInternalBalanceTransferRequest(params, opts) {
        return this._executeAction({ action: "accountInternalBalanceTransferRequest", params }, opts?.signal);
    }
    /**---Vault Methods---*/
    async depositToVault(params, opts) {
        return this._executeAction({ action: "depositToVault", params }, opts?.signal);
    }
    async redeemFromVault(params, opts) {
        return this._executeAction({ action: "redeemFromVault", params }, opts?.signal);
    }
    /**---Private Methods---*/
    async _executeAction(request, signal, execute = true) {
        const { action, params } = request;
        params.nonce = params.nonce ?? (await this.nonce());
        let signature = await signAction({
            wallet: this.wallet,
            action: params,
            txType: EXCHANGE_OP_CODES[action],
        }, {
            isTestnet: this.transport.isTestnet,
        });
        if (execute) {
            const response = await this.transport.request("exchange", {
                action: {
                    data: params,
                    type: String(EXCHANGE_OP_CODES[action]),
                },
                signature,
                nonce: params.nonce,
            }, signal);
            return response;
        }
        return { params, signature };
    }
}
