// @hotstuff-labs/ts-sdk/dist/apis/info.api.js
export class InfoClient {
    constructor(args) {
        this.transport = args.transport;
    }
    /**---Global Info Endpoints---*/
    async oracle(params, signal) {
        const request = {
            method: 'oracle',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async supportedCollateral(params, signal) {
        const request = {
            method: 'supported_collateral',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async instruments(params, signal) {
        const request = {
            method: 'instruments',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async ticker(params, signal) {
        const request = {
            method: 'ticker',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async orderbook(params, signal) {
        const request = {
            method: 'orderbook',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async trades(params, signal) {
        const request = {
            method: 'trades',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async mids(params, signal) {
        const request = {
            method: 'mids',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async bbo(params, signal) {
        const request = {
            method: 'bbo',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async chart(params, signal) {
        const request = {
            method: 'chart',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    /**Account Info Endpoints */
    async accountHistory(params, signal) {
        const request = {
            method: 'account_history',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async accountInfo(params, signal) {
        const request = {
            method: 'account_info',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async accountSummary(params, signal) {
        const request = {
            method: 'account_summary',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async openOrders(params, signal) {
        const request = {
            method: 'open_orders',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async positions(params, signal) {
        const request = {
            method: 'positions',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async orderHistory(params, signal) {
        const request = {
            method: 'order_history',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async fills(params, signal) {
        const request = {
            method: 'fills',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async fundingHistory(params, signal) {
        const request = {
            method: 'funding_history',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async transferHistory(params, signal) {
        const request = {
            method: 'transfer_history',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async instrumentLeverage(params, signal) {
        const request = {
            method: 'instrument_leverage',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async allAgents(params, signal) {
        const request = {
            method: 'all_agents',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async brokersCheck(params, signal) {
        const request = {
            method: 'brokers_check',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async referralSummary(params, signal) {
        const request = {
            method: 'referral_summary',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async userFeeInfo(params, signal) {
        const request = {
            method: 'user_fees',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    /**---Vault Info Endpoints---*/
    async vaults(params, signal) {
        const request = {
            method: 'vaults',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async subVaults(params, signal) {
        const request = {
            method: 'sub_vaults',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async vaultBalances(params, signal) {
        const request = {
            method: 'vault_balance',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    /** Explorer Info Endpoints */
    async blocks(params, signal) {
        const request = {
            method: 'blocks',
            params,
        };
        return this.transport.request('explorer', request, signal);
    }
    async blockDetails(params, signal) {
        const request = {
            method: 'block',
            params,
        };
        return this.transport.request('explorer', request, signal);
    }
    async transactions(params, signal) {
        const request = {
            method: 'transactions',
            params,
        };
        return this.transport.request('explorer', request, signal);
    }
    async transactionDetails(params, signal) {
        const request = {
            method: 'transaction',
            params,
        };
        return this.transport.request('explorer', request, signal);
    }
}
