// @hotstuff-labs/ts-sdk/dist/apis/info.api.js
export class InfoClient {
    constructor(args) {
        this.transport = args.transport;
    }
    /**---Global Info Endpoints---*/
    async oracle(params, signal) {
        const request = {
            method: 'oracle',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async supportedCollateral(params, signal) {
        const request = {
            method: 'supportedCollateral',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async instruments(params, signal) {
        const request = {
            method: 'instruments',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async ticker(params, signal) {
        const request = {
            method: 'ticker',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async orderbook(params, signal) {
        const request = {
            method: 'orderbook',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async trades(params, signal) {
        const request = {
            method: 'trades',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async mids(params, signal) {
        const request = {
            method: 'mids',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async bbo(params, signal) {
        const request = {
            method: 'bbo',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async chart(params, signal) {
        const request = {
            method: 'chart',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    /**Account Info Endpoints */
    async openOrders(params, signal) {
        const request = {
            method: 'openOrders',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async positions(params, signal) {
        const request = {
            method: 'positions',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async accountSummary(params, signal) {
        const request = {
            method: 'accountSummary',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async referralSummary(params, signal) {
        const request = {
            method: 'referralSummary',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async userFeeInfo(params, signal) {
        const request = {
            method: 'userFees',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async accountHistory(params, signal) {
        const request = {
            method: 'accountHistory',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async orderHistory(params, signal) {
        const request = {
            method: 'orderHistory',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async tradeHistory(params, signal) {
        const request = {
            method: 'fills',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async fundingHistory(params, signal) {
        const request = {
            method: 'fundingHistory',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async transferHistory(params, signal) {
        const request = {
            method: 'transferHistory',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async instrumentLeverage(params, signal) {
        const request = {
            method: 'instrumentLeverage',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async agents(params, signal) {
        const request = {
            method: 'allAgents',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async accountInfo(params, signal) {
        const request = {
            method: 'accountInfo',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    /**---Vault Info Endpoints---*/
    async vaults(params, signal) {
        const request = {
            method: 'vaults',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async subVaults(params, signal) {
        const request = {
            method: 'subVaults',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    async vaultBalances(params, signal) {
        const request = {
            method: 'vaultBalance',
            params,
        };
        return this.transport.request('info', request, signal);
    }
    /** Explorer Info Endpoints */
    async blocks(params, signal) {
        const request = {
            method: 'blocks',
            params,
        };
        return this.transport.request('explorer', request, signal);
    }
    async blockDetails(params, signal) {
        const request = {
            method: 'block',
            params,
        };
        return this.transport.request('explorer', request, signal);
    }
    async transactions(params, signal) {
        const request = {
            method: 'transactions',
            params,
        };
        return this.transport.request('explorer', request, signal);
    }
    async transactionDetails(params, signal) {
        const request = {
            method: 'transaction',
            params,
        };
        return this.transport.request('explorer', request, signal);
    }
}
