// @hotstuff-labs/ts-sdk/dist/apis/subscription.api.js
export class SubscriptionClient {
    constructor(args) {
        this.transport = args.transport;
    }
    /** Global Subscription Endpoints */
    async ticker(params, listener) {
        const result = await this.transport.subscribe('ticker', params, listener);
        const resultWithId = result;
        return {
            subscriptionId: resultWithId.subscriptionId,
            unsubscribe: () => this.transport.unsubscribe(resultWithId.subscriptionId),
        };
    }
    async mids(params, listener) {
        const result = await this.transport.subscribe('mids', params, listener);
        const resultWithId = result;
        return {
            subscriptionId: resultWithId.subscriptionId,
            unsubscribe: () => this.transport.unsubscribe(resultWithId.subscriptionId),
        };
    }
    async bbo(params, listener) {
        const result = await this.transport.subscribe('bbo', params, listener);
        const resultWithId = result;
        return {
            subscriptionId: resultWithId.subscriptionId,
            unsubscribe: () => this.transport.unsubscribe(resultWithId.subscriptionId),
        };
    }
    async orderbook(params, listener) {
        const request = {
            instrumentId: params.instrumentId,
            symbol: params.instrumentId,
        };
        const result = await this.transport.subscribe('orderbook', request, listener);
        const resultWithId = result;
        return {
            subscriptionId: resultWithId.subscriptionId,
            unsubscribe: () => this.transport.unsubscribe(resultWithId.subscriptionId),
        };
    }
    async trade(params, listener) {
        const request = {
            instrumentId: params.instrumentId,
            symbol: params.instrumentId,
        };
        const result = await this.transport.subscribe('trade', request, listener);
        const resultWithId = result;
        return {
            subscriptionId: resultWithId.subscriptionId,
            unsubscribe: () => this.transport.unsubscribe(resultWithId.subscriptionId),
        };
    }
    async index(listener) {
        const result = await this.transport.subscribe('index', {}, listener);
        const resultWithId = result;
        return {
            subscriptionId: resultWithId.subscriptionId,
            unsubscribe: () => this.transport.unsubscribe(resultWithId.subscriptionId),
        };
    }
    async chart(params, listener) {
        const result = await this.transport.subscribe('chart', params, listener);
        const resultWithId = result;
        return {
            subscriptionId: resultWithId.subscriptionId,
            unsubscribe: () => this.transport.unsubscribe(resultWithId.subscriptionId),
        };
    }
    /** Account Subscription Endpoints */
    async accountOrderUpdates(params, listener) {
        const request = {
            address: params.address,
            user: params.address,
        };
        const result = await this.transport.subscribe('order', request, listener);
        const resultWithId = result;
        return {
            subscriptionId: resultWithId.subscriptionId,
            unsubscribe: () => this.transport.unsubscribe(resultWithId.subscriptionId),
        };
    }
    async accountBalanceUpdates(params, listener) {
        const request = {
            address: params.address,
            user: params.address,
        };
        const result = await this.transport.subscribe('balance', request, listener);
        const resultWithId = result;
        return {
            subscriptionId: resultWithId.subscriptionId,
            unsubscribe: () => this.transport.unsubscribe(resultWithId.subscriptionId),
        };
    }
    async positions(params, listener) {
        const request = {
            address: params.address,
            user: params.address,
        };
        const result = await this.transport.subscribe('position', request, listener);
        const resultWithId = result;
        return {
            subscriptionId: resultWithId.subscriptionId,
            unsubscribe: () => this.transport.unsubscribe(resultWithId.subscriptionId),
        };
    }
    async fills(params, listener) {
        const request = {
            address: params.address,
            user: params.address,
        };
        const result = await this.transport.subscribe('fills', request, listener);
        const resultWithId = result;
        return {
            subscriptionId: resultWithId.subscriptionId,
            unsubscribe: () => this.transport.unsubscribe(resultWithId.subscriptionId),
        };
    }
    async accountSummary(params, listener) {
        const request = {
            user: params.user,
        };
        const result = await this.transport.subscribe('account_summary', request, listener);
        const resultWithId = result;
        return {
            subscriptionId: resultWithId.subscriptionId,
            unsubscribe: () => this.transport.unsubscribe(resultWithId.subscriptionId),
        };
    }
    /** Explorer Subscription Endpoints */
    async blocks(params, listener) {
        const request = {};
        const result = await this.transport.subscribe('blocks', request, listener);
        const resultWithId = result;
        return {
            subscriptionId: resultWithId.subscriptionId,
            unsubscribe: () => this.transport.unsubscribe(resultWithId.subscriptionId),
        };
    }
    async transactions(params, listener) {
        const request = {};
        const result = await this.transport.subscribe('transactions', request, listener);
        const resultWithId = result;
        return {
            subscriptionId: resultWithId.subscriptionId,
            unsubscribe: () => this.transport.unsubscribe(resultWithId.subscriptionId),
        };
    }
}
