// @yash-dynamo/mcp11/dist/index.js
#!/usr/bin/env node
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __commonJS = (cb, mod) => function __require() {
  return mod || (0, cb[__getOwnPropNames(cb)[0]])((mod = { exports: {} }).exports, mod), mod.exports;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));

// node_modules/@msgpack/msgpack/dist.cjs/utils/utf8.cjs
var require_utf8 = __commonJS({
  "node_modules/@msgpack/msgpack/dist.cjs/utils/utf8.cjs"(exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.utf8Count = utf8Count;
    exports.utf8EncodeJs = utf8EncodeJs;
    exports.utf8EncodeTE = utf8EncodeTE;
    exports.utf8Encode = utf8Encode;
    exports.utf8DecodeJs = utf8DecodeJs;
    exports.utf8DecodeTD = utf8DecodeTD;
    exports.utf8Decode = utf8Decode;
    function utf8Count(str) {
      const strLength = str.length;
      let byteLength = 0;
      let pos = 0;
      while (pos < strLength) {
        let value = str.charCodeAt(pos++);
        if ((value & 4294967168) === 0) {
          byteLength++;
          continue;
        } else if ((value & 4294965248) === 0) {
          byteLength += 2;
        } else {
          if (value >= 55296 && value <= 56319) {
            if (pos < strLength) {
              const extra = str.charCodeAt(pos);
              if ((extra & 64512) === 56320) {
                ++pos;
                value = ((value & 1023) << 10) + (extra & 1023) + 65536;
              }
            }
          }
          if ((value & 4294901760) === 0) {
            byteLength += 3;
          } else {
            byteLength += 4;
          }
        }
      }
      return byteLength;
    }
    function utf8EncodeJs(str, output, outputOffset) {
      const strLength = str.length;
      let offset = outputOffset;
      let pos = 0;
      while (pos < strLength) {
        let value = str.charCodeAt(pos++);
        if ((value & 4294967168) === 0) {
          output[offset++] = value;
          continue;
        } else if ((value & 4294965248) === 0) {
          output[offset++] = value >> 6 & 31 | 192;
        } else {
          if (value >= 55296 && value <= 56319) {
            if (pos < strLength) {
              const extra = str.charCodeAt(pos);
              if ((extra & 64512) === 56320) {
                ++pos;
                value = ((value & 1023) << 10) + (extra & 1023) + 65536;
              }
            }
          }
          if ((value & 4294901760) === 0) {
            output[offset++] = value >> 12 & 15 | 224;
            output[offset++] = value >> 6 & 63 | 128;
          } else {
            output[offset++] = value >> 18 & 7 | 240;
            output[offset++] = value >> 12 & 63 | 128;
            output[offset++] = value >> 6 & 63 | 128;
          }
        }
        output[offset++] = value & 63 | 128;
      }
    }
    var sharedTextEncoder = new TextEncoder();
    var TEXT_ENCODER_THRESHOLD = 50;
    function utf8EncodeTE(str, output, outputOffset) {
      sharedTextEncoder.encodeInto(str, output.subarray(outputOffset));
    }
    function utf8Encode(str, output, outputOffset) {
      if (str.length > TEXT_ENCODER_THRESHOLD) {
        utf8EncodeTE(str, output, outputOffset);
      } else {
        utf8EncodeJs(str, output, outputOffset);
      }
    }
    var CHUNK_SIZE = 4096;
    function utf8DecodeJs(bytes, inputOffset, byteLength) {
      let offset = inputOffset;
      const end = offset + byteLength;
      const units = [];
      let result = "";
      while (offset < end) {
        const byte1 = bytes[offset++];
        if ((byte1 & 128) === 0) {
          units.push(byte1);
        } else if ((byte1 & 224) === 192) {
          const byte2 = bytes[offset++] & 63;
          units.push((byte1 & 31) << 6 | byte2);
        } else if ((byte1 & 240) === 224) {
          const byte2 = bytes[offset++] & 63;
          const byte3 = bytes[offset++] & 63;
          units.push((byte1 & 31) << 12 | byte2 << 6 | byte3);
        } else if ((byte1 & 248) === 240) {
          const byte2 = bytes[offset++] & 63;
          const byte3 = bytes[offset++] & 63;
          const byte4 = bytes[offset++] & 63;
          let unit = (byte1 & 7) << 18 | byte2 << 12 | byte3 << 6 | byte4;
          if (unit > 65535) {
            unit -= 65536;
            units.push(unit >>> 10 & 1023 | 55296);
            unit = 56320 | unit & 1023;
          }
          units.push(unit);
        } else {
          units.push(byte1);
        }
        if (units.length >= CHUNK_SIZE) {
          result += String.fromCharCode(...units);
          units.length = 0;
        }
      }
      if (units.length > 0) {
        result += String.fromCharCode(...units);
      }
      return result;
    }
    var sharedTextDecoder = new TextDecoder();
    var TEXT_DECODER_THRESHOLD = 200;
    function utf8DecodeTD(bytes, inputOffset, byteLength) {
      const stringBytes = bytes.subarray(inputOffset, inputOffset + byteLength);
      return sharedTextDecoder.decode(stringBytes);
    }
    function utf8Decode(bytes, inputOffset, byteLength) {
      if (byteLength > TEXT_DECODER_THRESHOLD) {
        return utf8DecodeTD(bytes, inputOffset, byteLength);
      } else {
        return utf8DecodeJs(bytes, inputOffset, byteLength);
      }
    }
  }
});

// node_modules/@msgpack/msgpack/dist.cjs/ExtData.cjs
var require_ExtData = __commonJS({
  "node_modules/@msgpack/msgpack/dist.cjs/ExtData.cjs"(exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.ExtData = void 0;
    var ExtData = class {
      type;
      data;
      constructor(type, data) {
        this.type = type;
        this.data = data;
      }
    };
    exports.ExtData = ExtData;
  }
});

// node_modules/@msgpack/msgpack/dist.cjs/DecodeError.cjs
var require_DecodeError = __commonJS({
  "node_modules/@msgpack/msgpack/dist.cjs/DecodeError.cjs"(exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.DecodeError = void 0;
    var DecodeError = class _DecodeError extends Error {
      constructor(message) {
        super(message);
        const proto = Object.create(_DecodeError.prototype);
        Object.setPrototypeOf(this, proto);
        Object.defineProperty(this, "name", {
          configurable: true,
          enumerable: false,
          value: _DecodeError.name
        });
      }
    };
    exports.DecodeError = DecodeError;
  }
});

// node_modules/@msgpack/msgpack/dist.cjs/utils/int.cjs
var require_int = __commonJS({
  "node_modules/@msgpack/msgpack/dist.cjs/utils/int.cjs"(exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.UINT32_MAX = void 0;
    exports.setUint64 = setUint64;
    exports.setInt64 = setInt64;
    exports.getInt64 = getInt64;
    exports.getUint64 = getUint64;
    exports.UINT32_MAX = 4294967295;
    function setUint64(view, offset, value) {
      const high = value / 4294967296;
      const low = value;
      view.setUint32(offset, high);
      view.setUint32(offset + 4, low);
    }
    function setInt64(view, offset, value) {
      const high = Math.floor(value / 4294967296);
      const low = value;
      view.setUint32(offset, high);
      view.setUint32(offset + 4, low);
    }
    function getInt64(view, offset) {
      const high = view.getInt32(offset);
      const low = view.getUint32(offset + 4);
      return high * 4294967296 + low;
    }
    function getUint64(view, offset) {
      const high = view.getUint32(offset);
      const low = view.getUint32(offset + 4);
      return high * 4294967296 + low;
    }
  }
});

// node_modules/@msgpack/msgpack/dist.cjs/timestamp.cjs
var require_timestamp = __commonJS({
  "node_modules/@msgpack/msgpack/dist.cjs/timestamp.cjs"(exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.timestampExtension = exports.EXT_TIMESTAMP = void 0;
    exports.encodeTimeSpecToTimestamp = encodeTimeSpecToTimestamp;
    exports.encodeDateToTimeSpec = encodeDateToTimeSpec;
    exports.encodeTimestampExtension = encodeTimestampExtension;
    exports.decodeTimestampToTimeSpec = decodeTimestampToTimeSpec;
    exports.decodeTimestampExtension = decodeTimestampExtension;
    var DecodeError_ts_1 = require_DecodeError();
    var int_ts_1 = require_int();
    exports.EXT_TIMESTAMP = -1;
    var TIMESTAMP32_MAX_SEC = 4294967296 - 1;
    var TIMESTAMP64_MAX_SEC = 17179869184 - 1;
    function encodeTimeSpecToTimestamp({ sec, nsec }) {
      if (sec >= 0 && nsec >= 0 && sec <= TIMESTAMP64_MAX_SEC) {
        if (nsec === 0 && sec <= TIMESTAMP32_MAX_SEC) {
          const rv = new Uint8Array(4);
          const view = new DataView(rv.buffer);
          view.setUint32(0, sec);
          return rv;
        } else {
          const secHigh = sec / 4294967296;
          const secLow = sec & 4294967295;
          const rv = new Uint8Array(8);
          const view = new DataView(rv.buffer);
          view.setUint32(0, nsec << 2 | secHigh & 3);
          view.setUint32(4, secLow);
          return rv;
        }
      } else {
        const rv = new Uint8Array(12);
        const view = new DataView(rv.buffer);
        view.setUint32(0, nsec);
        (0, int_ts_1.setInt64)(view, 4, sec);
        return rv;
      }
    }
    function encodeDateToTimeSpec(date) {
      const msec = date.getTime();
      const sec = Math.floor(msec / 1e3);
      const nsec = (msec - sec * 1e3) * 1e6;
      const nsecInSec = Math.floor(nsec / 1e9);
      return {
        sec: sec + nsecInSec,
        nsec: nsec - nsecInSec * 1e9
      };
    }
    function encodeTimestampExtension(object) {
      if (object instanceof Date) {
        const timeSpec = encodeDateToTimeSpec(object);
        return encodeTimeSpecToTimestamp(timeSpec);
      } else {
        return null;
      }
    }
    function decodeTimestampToTimeSpec(data) {
      const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
      switch (data.byteLength) {
        case 4: {
          const sec = view.getUint32(0);
          const nsec = 0;
          return { sec, nsec };
        }
        case 8: {
          const nsec30AndSecHigh2 = view.getUint32(0);
          const secLow32 = view.getUint32(4);
          const sec = (nsec30AndSecHigh2 & 3) * 4294967296 + secLow32;
          const nsec = nsec30AndSecHigh2 >>> 2;
          return { sec, nsec };
        }
        case 12: {
          const sec = (0, int_ts_1.getInt64)(view, 4);
          const nsec = view.getUint32(0);
          return { sec, nsec };
        }
        default:
          throw new DecodeError_ts_1.DecodeError(`Unrecognized data size for timestamp (expected 4, 8, or 12): ${data.length}`);
      }
    }
    function decodeTimestampExtension(data) {
      const timeSpec = decodeTimestampToTimeSpec(data);
      return new Date(timeSpec.sec * 1e3 + timeSpec.nsec / 1e6);
    }
    exports.timestampExtension = {
      type: exports.EXT_TIMESTAMP,
      encode: encodeTimestampExtension,
      decode: decodeTimestampExtension
    };
  }
});

// node_modules/@msgpack/msgpack/dist.cjs/ExtensionCodec.cjs
var require_ExtensionCodec = __commonJS({
  "node_modules/@msgpack/msgpack/dist.cjs/ExtensionCodec.cjs"(exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.ExtensionCodec = void 0;
    var ExtData_ts_1 = require_ExtData();
    var timestamp_ts_1 = require_timestamp();
    var ExtensionCodec = class _ExtensionCodec {
      static defaultCodec = new _ExtensionCodec();
      // ensures ExtensionCodecType<X> matches ExtensionCodec<X>
      // this will make type errors a lot more clear
      // eslint-disable-next-line @typescript-eslint/naming-convention
      __brand;
      // built-in extensions
      builtInEncoders = [];
      builtInDecoders = [];
      // custom extensions
      encoders = [];
      decoders = [];
      constructor() {
        this.register(timestamp_ts_1.timestampExtension);
      }
      register({ type, encode: encode2, decode }) {
        if (type >= 0) {
          this.encoders[type] = encode2;
          this.decoders[type] = decode;
        } else {
          const index = -1 - type;
          this.builtInEncoders[index] = encode2;
          this.builtInDecoders[index] = decode;
        }
      }
      tryToEncode(object, context) {
        for (let i = 0; i < this.builtInEncoders.length; i++) {
          const encodeExt = this.builtInEncoders[i];
          if (encodeExt != null) {
            const data = encodeExt(object, context);
            if (data != null) {
              const type = -1 - i;
              return new ExtData_ts_1.ExtData(type, data);
            }
          }
        }
        for (let i = 0; i < this.encoders.length; i++) {
          const encodeExt = this.encoders[i];
          if (encodeExt != null) {
            const data = encodeExt(object, context);
            if (data != null) {
              const type = i;
              return new ExtData_ts_1.ExtData(type, data);
            }
          }
        }
        if (object instanceof ExtData_ts_1.ExtData) {
          return object;
        }
        return null;
      }
      decode(data, type, context) {
        const decodeExt = type < 0 ? this.builtInDecoders[-1 - type] : this.decoders[type];
        if (decodeExt) {
          return decodeExt(data, type, context);
        } else {
          return new ExtData_ts_1.ExtData(type, data);
        }
      }
    };
    exports.ExtensionCodec = ExtensionCodec;
  }
});

// node_modules/@msgpack/msgpack/dist.cjs/utils/typedArrays.cjs
var require_typedArrays = __commonJS({
  "node_modules/@msgpack/msgpack/dist.cjs/utils/typedArrays.cjs"(exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.ensureUint8Array = ensureUint8Array;
    function isArrayBufferLike(buffer) {
      return buffer instanceof ArrayBuffer || typeof SharedArrayBuffer !== "undefined" && buffer instanceof SharedArrayBuffer;
    }
    function ensureUint8Array(buffer) {
      if (buffer instanceof Uint8Array) {
        return buffer;
      } else if (ArrayBuffer.isView(buffer)) {
        return new Uint8Array(buffer.buffer, buffer.byteOffset, buffer.byteLength);
      } else if (isArrayBufferLike(buffer)) {
        return new Uint8Array(buffer);
      } else {
        return Uint8Array.from(buffer);
      }
    }
  }
});

// node_modules/@msgpack/msgpack/dist.cjs/Encoder.cjs
var require_Encoder = __commonJS({
  "node_modules/@msgpack/msgpack/dist.cjs/Encoder.cjs"(exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.Encoder = exports.DEFAULT_INITIAL_BUFFER_SIZE = exports.DEFAULT_MAX_DEPTH = void 0;
    var utf8_ts_1 = require_utf8();
    var ExtensionCodec_ts_1 = require_ExtensionCodec();
    var int_ts_1 = require_int();
    var typedArrays_ts_1 = require_typedArrays();
    exports.DEFAULT_MAX_DEPTH = 100;
    exports.DEFAULT_INITIAL_BUFFER_SIZE = 2048;
    var Encoder = class _Encoder {
      extensionCodec;
      context;
      useBigInt64;
      maxDepth;
      initialBufferSize;
      sortKeys;
      forceFloat32;
      ignoreUndefined;
      forceIntegerToFloat;
      pos;
      view;
      bytes;
      entered = false;
      constructor(options) {
        this.extensionCodec = options?.extensionCodec ?? ExtensionCodec_ts_1.ExtensionCodec.defaultCodec;
        this.context = options?.context;
        this.useBigInt64 = options?.useBigInt64 ?? false;
        this.maxDepth = options?.maxDepth ?? exports.DEFAULT_MAX_DEPTH;
        this.initialBufferSize = options?.initialBufferSize ?? exports.DEFAULT_INITIAL_BUFFER_SIZE;
        this.sortKeys = options?.sortKeys ?? false;
        this.forceFloat32 = options?.forceFloat32 ?? false;
        this.ignoreUndefined = options?.ignoreUndefined ?? false;
        this.forceIntegerToFloat = options?.forceIntegerToFloat ?? false;
        this.pos = 0;
        this.view = new DataView(new ArrayBuffer(this.initialBufferSize));
        this.bytes = new Uint8Array(this.view.buffer);
      }
      clone() {
        return new _Encoder({
          extensionCodec: this.extensionCodec,
          context: this.context,
          useBigInt64: this.useBigInt64,
          maxDepth: this.maxDepth,
          initialBufferSize: this.initialBufferSize,
          sortKeys: this.sortKeys,
          forceFloat32: this.forceFloat32,
          ignoreUndefined: this.ignoreUndefined,
          forceIntegerToFloat: this.forceIntegerToFloat
        });
      }
      reinitializeState() {
        this.pos = 0;
      }
      /**
       * This is almost equivalent to {@link Encoder#encode}, but it returns an reference of the encoder's internal buffer and thus much faster than {@link Encoder#encode}.
       *
       * @returns Encodes the object and returns a shared reference the encoder's internal buffer.
       */
      encodeSharedRef(object) {
        if (this.entered) {
          const instance = this.clone();
          return instance.encodeSharedRef(object);
        }
        try {
          this.entered = true;
          this.reinitializeState();
          this.doEncode(object, 1);
          return this.bytes.subarray(0, this.pos);
        } finally {
          this.entered = false;
        }
      }
      /**
       * @returns Encodes the object and returns a copy of the encoder's internal buffer.
       */
      encode(object) {
        if (this.entered) {
          const instance = this.clone();
          return instance.encode(object);
        }
        try {
          this.entered = true;
          this.reinitializeState();
          this.doEncode(object, 1);
          return this.bytes.slice(0, this.pos);
        } finally {
          this.entered = false;
        }
      }
      doEncode(object, depth) {
        if (depth > this.maxDepth) {
          throw new Error(`Too deep objects in depth ${depth}`);
        }
        if (object == null) {
          this.encodeNil();
        } else if (typeof object === "boolean") {
          this.encodeBoolean(object);
        } else if (typeof object === "number") {
          if (!this.forceIntegerToFloat) {
            this.encodeNumber(object);
          } else {
            this.encodeNumberAsFloat(object);
          }
        } else if (typeof object === "string") {
          this.encodeString(object);
        } else if (this.useBigInt64 && typeof object === "bigint") {
          this.encodeBigInt64(object);
        } else {
          this.encodeObject(object, depth);
        }
      }
      ensureBufferSizeToWrite(sizeToWrite) {
        const requiredSize = this.pos + sizeToWrite;
        if (this.view.byteLength < requiredSize) {
          this.resizeBuffer(requiredSize * 2);
        }
      }
      resizeBuffer(newSize) {
        const newBuffer = new ArrayBuffer(newSize);
        const newBytes = new Uint8Array(newBuffer);
        const newView = new DataView(newBuffer);
        newBytes.set(this.bytes);
        this.view = newView;
        this.bytes = newBytes;
      }
      encodeNil() {
        this.writeU8(192);
      }
      encodeBoolean(object) {
        if (object === false) {
          this.writeU8(194);
        } else {
          this.writeU8(195);
        }
      }
      encodeNumber(object) {
        if (!this.forceIntegerToFloat && Number.isSafeInteger(object)) {
          if (object >= 0) {
            if (object < 128) {
              this.writeU8(object);
            } else if (object < 256) {
              this.writeU8(204);
              this.writeU8(object);
            } else if (object < 65536) {
              this.writeU8(205);
              this.writeU16(object);
            } else if (object < 4294967296) {
              this.writeU8(206);
              this.writeU32(object);
            } else if (!this.useBigInt64) {
              this.writeU8(207);
              this.writeU64(object);
            } else {
              this.encodeNumberAsFloat(object);
            }
          } else {
            if (object >= -32) {
              this.writeU8(224 | object + 32);
            } else if (object >= -128) {
              this.writeU8(208);
              this.writeI8(object);
            } else if (object >= -32768) {
              this.writeU8(209);
              this.writeI16(object);
            } else if (object >= -2147483648) {
              this.writeU8(210);
              this.writeI32(object);
            } else if (!this.useBigInt64) {
              this.writeU8(211);
              this.writeI64(object);
            } else {
              this.encodeNumberAsFloat(object);
            }
          }
        } else {
          this.encodeNumberAsFloat(object);
        }
      }
      encodeNumberAsFloat(object) {
        if (this.forceFloat32) {
          this.writeU8(202);
          this.writeF32(object);
        } else {
          this.writeU8(203);
          this.writeF64(object);
        }
      }
      encodeBigInt64(object) {
        if (object >= BigInt(0)) {
          this.writeU8(207);
          this.writeBigUint64(object);
        } else {
          this.writeU8(211);
          this.writeBigInt64(object);
        }
      }
      writeStringHeader(byteLength) {
        if (byteLength < 32) {
          this.writeU8(160 + byteLength);
        } else if (byteLength < 256) {
          this.writeU8(217);
          this.writeU8(byteLength);
        } else if (byteLength < 65536) {
          this.writeU8(218);
          this.writeU16(byteLength);
        } else if (byteLength < 4294967296) {
          this.writeU8(219);
          this.writeU32(byteLength);
        } else {
          throw new Error(`Too long string: ${byteLength} bytes in UTF-8`);
        }
      }
      encodeString(object) {
        const maxHeaderSize = 1 + 4;
        const byteLength = (0, utf8_ts_1.utf8Count)(object);
        this.ensureBufferSizeToWrite(maxHeaderSize + byteLength);
        this.writeStringHeader(byteLength);
        (0, utf8_ts_1.utf8Encode)(object, this.bytes, this.pos);
        this.pos += byteLength;
      }
      encodeObject(object, depth) {
        const ext = this.extensionCodec.tryToEncode(object, this.context);
        if (ext != null) {
          this.encodeExtension(ext);
        } else if (Array.isArray(object)) {
          this.encodeArray(object, depth);
        } else if (ArrayBuffer.isView(object)) {
          this.encodeBinary(object);
        } else if (typeof object === "object") {
          this.encodeMap(object, depth);
        } else {
          throw new Error(`Unrecognized object: ${Object.prototype.toString.apply(object)}`);
        }
      }
      encodeBinary(object) {
        const size = object.byteLength;
        if (size < 256) {
          this.writeU8(196);
          this.writeU8(size);
        } else if (size < 65536) {
          this.writeU8(197);
          this.writeU16(size);
        } else if (size < 4294967296) {
          this.writeU8(198);
          this.writeU32(size);
        } else {
          throw new Error(`Too large binary: ${size}`);
        }
        const bytes = (0, typedArrays_ts_1.ensureUint8Array)(object);
        this.writeU8a(bytes);
      }
      encodeArray(object, depth) {
        const size = object.length;
        if (size < 16) {
          this.writeU8(144 + size);
        } else if (size < 65536) {
          this.writeU8(220);
          this.writeU16(size);
        } else if (size < 4294967296) {
          this.writeU8(221);
          this.writeU32(size);
        } else {
          throw new Error(`Too large array: ${size}`);
        }
        for (const item of object) {
          this.doEncode(item, depth + 1);
        }
      }
      countWithoutUndefined(object, keys) {
        let count = 0;
        for (const key of keys) {
          if (object[key] !== void 0) {
            count++;
          }
        }
        return count;
      }
      encodeMap(object, depth) {
        const keys = Object.keys(object);
        if (this.sortKeys) {
          keys.sort();
        }
        const size = this.ignoreUndefined ? this.countWithoutUndefined(object, keys) : keys.length;
        if (size < 16) {
          this.writeU8(128 + size);
        } else if (size < 65536) {
          this.writeU8(222);
          this.writeU16(size);
        } else if (size < 4294967296) {
          this.writeU8(223);
          this.writeU32(size);
        } else {
          throw new Error(`Too large map object: ${size}`);
        }
        for (const key of keys) {
          const value = object[key];
          if (!(this.ignoreUndefined && value === void 0)) {
            this.encodeString(key);
            this.doEncode(value, depth + 1);
          }
        }
      }
      encodeExtension(ext) {
        if (typeof ext.data === "function") {
          const data = ext.data(this.pos + 6);
          const size2 = data.length;
          if (size2 >= 4294967296) {
            throw new Error(`Too large extension object: ${size2}`);
          }
          this.writeU8(201);
          this.writeU32(size2);
          this.writeI8(ext.type);
          this.writeU8a(data);
          return;
        }
        const size = ext.data.length;
        if (size === 1) {
          this.writeU8(212);
        } else if (size === 2) {
          this.writeU8(213);
        } else if (size === 4) {
          this.writeU8(214);
        } else if (size === 8) {
          this.writeU8(215);
        } else if (size === 16) {
          this.writeU8(216);
        } else if (size < 256) {
          this.writeU8(199);
          this.writeU8(size);
        } else if (size < 65536) {
          this.writeU8(200);
          this.writeU16(size);
        } else if (size < 4294967296) {
          this.writeU8(201);
          this.writeU32(size);
        } else {
          throw new Error(`Too large extension object: ${size}`);
        }
        this.writeI8(ext.type);
        this.writeU8a(ext.data);
      }
      writeU8(value) {
        this.ensureBufferSizeToWrite(1);
        this.view.setUint8(this.pos, value);
        this.pos++;
      }
      writeU8a(values) {
        const size = values.length;
        this.ensureBufferSizeToWrite(size);
        this.bytes.set(values, this.pos);
        this.pos += size;
      }
      writeI8(value) {
        this.ensureBufferSizeToWrite(1);
        this.view.setInt8(this.pos, value);
        this.pos++;
      }
      writeU16(value) {
        this.ensureBufferSizeToWrite(2);
        this.view.setUint16(this.pos, value);
        this.pos += 2;
      }
      writeI16(value) {
        this.ensureBufferSizeToWrite(2);
        this.view.setInt16(this.pos, value);
        this.pos += 2;
      }
      writeU32(value) {
        this.ensureBufferSizeToWrite(4);
        this.view.setUint32(this.pos, value);
        this.pos += 4;
      }
      writeI32(value) {
        this.ensureBufferSizeToWrite(4);
        this.view.setInt32(this.pos, value);
        this.pos += 4;
      }
      writeF32(value) {
        this.ensureBufferSizeToWrite(4);
        this.view.setFloat32(this.pos, value);
        this.pos += 4;
      }
      writeF64(value) {
        this.ensureBufferSizeToWrite(8);
        this.view.setFloat64(this.pos, value);
        this.pos += 8;
      }
      writeU64(value) {
        this.ensureBufferSizeToWrite(8);
        (0, int_ts_1.setUint64)(this.view, this.pos, value);
        this.pos += 8;
      }
      writeI64(value) {
        this.ensureBufferSizeToWrite(8);
        (0, int_ts_1.setInt64)(this.view, this.pos, value);
        this.pos += 8;
      }
      writeBigUint64(value) {
        this.ensureBufferSizeToWrite(8);
        this.view.setBigUint64(this.pos, value);
        this.pos += 8;
      }
      writeBigInt64(value) {
        this.ensureBufferSizeToWrite(8);
        this.view.setBigInt64(this.pos, value);
        this.pos += 8;
      }
    };
    exports.Encoder = Encoder;
  }
});

// node_modules/@msgpack/msgpack/dist.cjs/encode.cjs
var require_encode = __commonJS({
  "node_modules/@msgpack/msgpack/dist.cjs/encode.cjs"(exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.encode = encode2;
    var Encoder_ts_1 = require_Encoder();
    function encode2(value, options) {
      const encoder = new Encoder_ts_1.Encoder(options);
      return encoder.encodeSharedRef(value);
    }
  }
});

// node_modules/@msgpack/msgpack/dist.cjs/utils/prettyByte.cjs
var require_prettyByte = __commonJS({
  "node_modules/@msgpack/msgpack/dist.cjs/utils/prettyByte.cjs"(exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.prettyByte = prettyByte;
    function prettyByte(byte) {
      return `${byte < 0 ? "-" : ""}0x${Math.abs(byte).toString(16).padStart(2, "0")}`;
    }
  }
});

// node_modules/@msgpack/msgpack/dist.cjs/CachedKeyDecoder.cjs
var require_CachedKeyDecoder = __commonJS({
  "node_modules/@msgpack/msgpack/dist.cjs/CachedKeyDecoder.cjs"(exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.CachedKeyDecoder = void 0;
    var utf8_ts_1 = require_utf8();
    var DEFAULT_MAX_KEY_LENGTH = 16;
    var DEFAULT_MAX_LENGTH_PER_KEY = 16;
    var CachedKeyDecoder = class {
      hit = 0;
      miss = 0;
      caches;
      maxKeyLength;
      maxLengthPerKey;
      constructor(maxKeyLength = DEFAULT_MAX_KEY_LENGTH, maxLengthPerKey = DEFAULT_MAX_LENGTH_PER_KEY) {
        this.maxKeyLength = maxKeyLength;
        this.maxLengthPerKey = maxLengthPerKey;
        this.caches = [];
        for (let i = 0; i < this.maxKeyLength; i++) {
          this.caches.push([]);
        }
      }
      canBeCached(byteLength) {
        return byteLength > 0 && byteLength <= this.maxKeyLength;
      }
      find(bytes, inputOffset, byteLength) {
        const records = this.caches[byteLength - 1];
        FIND_CHUNK: for (const record of records) {
          const recordBytes = record.bytes;
          for (let j = 0; j < byteLength; j++) {
            if (recordBytes[j] !== bytes[inputOffset + j]) {
              continue FIND_CHUNK;
            }
          }
          return record.str;
        }
        return null;
      }
      store(bytes, value) {
        const records = this.caches[bytes.length - 1];
        const record = { bytes, str: value };
        if (records.length >= this.maxLengthPerKey) {
          records[Math.random() * records.length | 0] = record;
        } else {
          records.push(record);
        }
      }
      decode(bytes, inputOffset, byteLength) {
        const cachedValue = this.find(bytes, inputOffset, byteLength);
        if (cachedValue != null) {
          this.hit++;
          return cachedValue;
        }
        this.miss++;
        const str = (0, utf8_ts_1.utf8DecodeJs)(bytes, inputOffset, byteLength);
        const slicedCopyOfBytes = Uint8Array.prototype.slice.call(bytes, inputOffset, inputOffset + byteLength);
        this.store(slicedCopyOfBytes, str);
        return str;
      }
    };
    exports.CachedKeyDecoder = CachedKeyDecoder;
  }
});

// node_modules/@msgpack/msgpack/dist.cjs/Decoder.cjs
var require_Decoder = __commonJS({
  "node_modules/@msgpack/msgpack/dist.cjs/Decoder.cjs"(exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.Decoder = void 0;
    var prettyByte_ts_1 = require_prettyByte();
    var ExtensionCodec_ts_1 = require_ExtensionCodec();
    var int_ts_1 = require_int();
    var utf8_ts_1 = require_utf8();
    var typedArrays_ts_1 = require_typedArrays();
    var CachedKeyDecoder_ts_1 = require_CachedKeyDecoder();
    var DecodeError_ts_1 = require_DecodeError();
    var STATE_ARRAY = "array";
    var STATE_MAP_KEY = "map_key";
    var STATE_MAP_VALUE = "map_value";
    var mapKeyConverter = (key) => {
      if (typeof key === "string" || typeof key === "number") {
        return key;
      }
      throw new DecodeError_ts_1.DecodeError("The type of key must be string or number but " + typeof key);
    };
    var StackPool = class {
      stack = [];
      stackHeadPosition = -1;
      get length() {
        return this.stackHeadPosition + 1;
      }
      top() {
        return this.stack[this.stackHeadPosition];
      }
      pushArrayState(size) {
        const state = this.getUninitializedStateFromPool();
        state.type = STATE_ARRAY;
        state.position = 0;
        state.size = size;
        state.array = new Array(size);
      }
      pushMapState(size) {
        const state = this.getUninitializedStateFromPool();
        state.type = STATE_MAP_KEY;
        state.readCount = 0;
        state.size = size;
        state.map = {};
      }
      getUninitializedStateFromPool() {
        this.stackHeadPosition++;
        if (this.stackHeadPosition === this.stack.length) {
          const partialState = {
            type: void 0,
            size: 0,
            array: void 0,
            position: 0,
            readCount: 0,
            map: void 0,
            key: null
          };
          this.stack.push(partialState);
        }
        return this.stack[this.stackHeadPosition];
      }
      release(state) {
        const topStackState = this.stack[this.stackHeadPosition];
        if (topStackState !== state) {
          throw new Error("Invalid stack state. Released state is not on top of the stack.");
        }
        if (state.type === STATE_ARRAY) {
          const partialState = state;
          partialState.size = 0;
          partialState.array = void 0;
          partialState.position = 0;
          partialState.type = void 0;
        }
        if (state.type === STATE_MAP_KEY || state.type === STATE_MAP_VALUE) {
          const partialState = state;
          partialState.size = 0;
          partialState.map = void 0;
          partialState.readCount = 0;
          partialState.type = void 0;
        }
        this.stackHeadPosition--;
      }
      reset() {
        this.stack.length = 0;
        this.stackHeadPosition = -1;
      }
    };
    var HEAD_BYTE_REQUIRED = -1;
    var EMPTY_VIEW = new DataView(new ArrayBuffer(0));
    var EMPTY_BYTES = new Uint8Array(EMPTY_VIEW.buffer);
    try {
      EMPTY_VIEW.getInt8(0);
    } catch (e) {
      if (!(e instanceof RangeError)) {
        throw new Error("This module is not supported in the current JavaScript engine because DataView does not throw RangeError on out-of-bounds access");
      }
    }
    var MORE_DATA = new RangeError("Insufficient data");
    var sharedCachedKeyDecoder = new CachedKeyDecoder_ts_1.CachedKeyDecoder();
    var Decoder = class _Decoder {
      extensionCodec;
      context;
      useBigInt64;
      rawStrings;
      maxStrLength;
      maxBinLength;
      maxArrayLength;
      maxMapLength;
      maxExtLength;
      keyDecoder;
      mapKeyConverter;
      totalPos = 0;
      pos = 0;
      view = EMPTY_VIEW;
      bytes = EMPTY_BYTES;
      headByte = HEAD_BYTE_REQUIRED;
      stack = new StackPool();
      entered = false;
      constructor(options) {
        this.extensionCodec = options?.extensionCodec ?? ExtensionCodec_ts_1.ExtensionCodec.defaultCodec;
        this.context = options?.context;
        this.useBigInt64 = options?.useBigInt64 ?? false;
        this.rawStrings = options?.rawStrings ?? false;
        this.maxStrLength = options?.maxStrLength ?? int_ts_1.UINT32_MAX;
        this.maxBinLength = options?.maxBinLength ?? int_ts_1.UINT32_MAX;
        this.maxArrayLength = options?.maxArrayLength ?? int_ts_1.UINT32_MAX;
        this.maxMapLength = options?.maxMapLength ?? int_ts_1.UINT32_MAX;
        this.maxExtLength = options?.maxExtLength ?? int_ts_1.UINT32_MAX;
        this.keyDecoder = options?.keyDecoder !== void 0 ? options.keyDecoder : sharedCachedKeyDecoder;
        this.mapKeyConverter = options?.mapKeyConverter ?? mapKeyConverter;
      }
      clone() {
        return new _Decoder({
          extensionCodec: this.extensionCodec,
          context: this.context,
          useBigInt64: this.useBigInt64,
          rawStrings: this.rawStrings,
          maxStrLength: this.maxStrLength,
          maxBinLength: this.maxBinLength,
          maxArrayLength: this.maxArrayLength,
          maxMapLength: this.maxMapLength,
          maxExtLength: this.maxExtLength,
          keyDecoder: this.keyDecoder
        });
      }
      reinitializeState() {
        this.totalPos = 0;
        this.headByte = HEAD_BYTE_REQUIRED;
        this.stack.reset();
      }
      setBuffer(buffer) {
        const bytes = (0, typedArrays_ts_1.ensureUint8Array)(buffer);
        this.bytes = bytes;
        this.view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
        this.pos = 0;
      }
      appendBuffer(buffer) {
        if (this.headByte === HEAD_BYTE_REQUIRED && !this.hasRemaining(1)) {
          this.setBuffer(buffer);
        } else {
          const remainingData = this.bytes.subarray(this.pos);
          const newData = (0, typedArrays_ts_1.ensureUint8Array)(buffer);
          const newBuffer = new Uint8Array(remainingData.length + newData.length);
          newBuffer.set(remainingData);
          newBuffer.set(newData, remainingData.length);
          this.setBuffer(newBuffer);
        }
      }
      hasRemaining(size) {
        return this.view.byteLength - this.pos >= size;
      }
      createExtraByteError(posToShow) {
        const { view, pos } = this;
        return new RangeError(`Extra ${view.byteLength - pos} of ${view.byteLength} byte(s) found at buffer[${posToShow}]`);
      }
      /**
       * @throws {@link DecodeError}
       * @throws {@link RangeError}
       */
      decode(buffer) {
        if (this.entered) {
          const instance = this.clone();
          return instance.decode(buffer);
        }
        try {
          this.entered = true;
          this.reinitializeState();
          this.setBuffer(buffer);
          const object = this.doDecodeSync();
          if (this.hasRemaining(1)) {
            throw this.createExtraByteError(this.pos);
          }
          return object;
        } finally {
          this.entered = false;
        }
      }
      *decodeMulti(buffer) {
        if (this.entered) {
          const instance = this.clone();
          yield* instance.decodeMulti(buffer);
          return;
        }
        try {
          this.entered = true;
          this.reinitializeState();
          this.setBuffer(buffer);
          while (this.hasRemaining(1)) {
            yield this.doDecodeSync();
          }
        } finally {
          this.entered = false;
        }
      }
      async decodeAsync(stream) {
        if (this.entered) {
          const instance = this.clone();
          return instance.decodeAsync(stream);
        }
        try {
          this.entered = true;
          let decoded = false;
          let object;
          for await (const buffer of stream) {
            if (decoded) {
              this.entered = false;
              throw this.createExtraByteError(this.totalPos);
            }
            this.appendBuffer(buffer);
            try {
              object = this.doDecodeSync();
              decoded = true;
            } catch (e) {
              if (!(e instanceof RangeError)) {
                throw e;
              }
            }
            this.totalPos += this.pos;
          }
          if (decoded) {
            if (this.hasRemaining(1)) {
              throw this.createExtraByteError(this.totalPos);
            }
            return object;
          }
          const { headByte, pos, totalPos } = this;
          throw new RangeError(`Insufficient data in parsing ${(0, prettyByte_ts_1.prettyByte)(headByte)} at ${totalPos} (${pos} in the current buffer)`);
        } finally {
          this.entered = false;
        }
      }
      decodeArrayStream(stream) {
        return this.decodeMultiAsync(stream, true);
      }
      decodeStream(stream) {
        return this.decodeMultiAsync(stream, false);
      }
      async *decodeMultiAsync(stream, isArray) {
        if (this.entered) {
          const instance = this.clone();
          yield* instance.decodeMultiAsync(stream, isArray);
          return;
        }
        try {
          this.entered = true;
          let isArrayHeaderRequired = isArray;
          let arrayItemsLeft = -1;
          for await (const buffer of stream) {
            if (isArray && arrayItemsLeft === 0) {
              throw this.createExtraByteError(this.totalPos);
            }
            this.appendBuffer(buffer);
            if (isArrayHeaderRequired) {
              arrayItemsLeft = this.readArraySize();
              isArrayHeaderRequired = false;
              this.complete();
            }
            try {
              while (true) {
                yield this.doDecodeSync();
                if (--arrayItemsLeft === 0) {
                  break;
                }
              }
            } catch (e) {
              if (!(e instanceof RangeError)) {
                throw e;
              }
            }
            this.totalPos += this.pos;
          }
        } finally {
          this.entered = false;
        }
      }
      doDecodeSync() {
        DECODE: while (true) {
          const headByte = this.readHeadByte();
          let object;
          if (headByte >= 224) {
            object = headByte - 256;
          } else if (headByte < 192) {
            if (headByte < 128) {
              object = headByte;
            } else if (headByte < 144) {
              const size = headByte - 128;
              if (size !== 0) {
                this.pushMapState(size);
                this.complete();
                continue DECODE;
              } else {
                object = {};
              }
            } else if (headByte < 160) {
              const size = headByte - 144;
              if (size !== 0) {
                this.pushArrayState(size);
                this.complete();
                continue DECODE;
              } else {
                object = [];
              }
            } else {
              const byteLength = headByte - 160;
              object = this.decodeString(byteLength, 0);
            }
          } else if (headByte === 192) {
            object = null;
          } else if (headByte === 194) {
            object = false;
          } else if (headByte === 195) {
            object = true;
          } else if (headByte === 202) {
            object = this.readF32();
          } else if (headByte === 203) {
            object = this.readF64();
          } else if (headByte === 204) {
            object = this.readU8();
          } else if (headByte === 205) {
            object = this.readU16();
          } else if (headByte === 206) {
            object = this.readU32();
          } else if (headByte === 207) {
            if (this.useBigInt64) {
              object = this.readU64AsBigInt();
            } else {
              object = this.readU64();
            }
          } else if (headByte === 208) {
            object = this.readI8();
          } else if (headByte === 209) {
            object = this.readI16();
          } else if (headByte === 210) {
            object = this.readI32();
          } else if (headByte === 211) {
            if (this.useBigInt64) {
              object = this.readI64AsBigInt();
            } else {
              object = this.readI64();
            }
          } else if (headByte === 217) {
            const byteLength = this.lookU8();
            object = this.decodeString(byteLength, 1);
          } else if (headByte === 218) {
            const byteLength = this.lookU16();
            object = this.decodeString(byteLength, 2);
          } else if (headByte === 219) {
            const byteLength = this.lookU32();
            object = this.decodeString(byteLength, 4);
          } else if (headByte === 220) {
            const size = this.readU16();
            if (size !== 0) {
              this.pushArrayState(size);
              this.complete();
              continue DECODE;
            } else {
              object = [];
            }
          } else if (headByte === 221) {
            const size = this.readU32();
            if (size !== 0) {
              this.pushArrayState(size);
              this.complete();
              continue DECODE;
            } else {
              object = [];
            }
          } else if (headByte === 222) {
            const size = this.readU16();
            if (size !== 0) {
              this.pushMapState(size);
              this.complete();
              continue DECODE;
            } else {
              object = {};
            }
          } else if (headByte === 223) {
            const size = this.readU32();
            if (size !== 0) {
              this.pushMapState(size);
              this.complete();
              continue DECODE;
            } else {
              object = {};
            }
          } else if (headByte === 196) {
            const size = this.lookU8();
            object = this.decodeBinary(size, 1);
          } else if (headByte === 197) {
            const size = this.lookU16();
            object = this.decodeBinary(size, 2);
          } else if (headByte === 198) {
            const size = this.lookU32();
            object = this.decodeBinary(size, 4);
          } else if (headByte === 212) {
            object = this.decodeExtension(1, 0);
          } else if (headByte === 213) {
            object = this.decodeExtension(2, 0);
          } else if (headByte === 214) {
            object = this.decodeExtension(4, 0);
          } else if (headByte === 215) {
            object = this.decodeExtension(8, 0);
          } else if (headByte === 216) {
            object = this.decodeExtension(16, 0);
          } else if (headByte === 199) {
            const size = this.lookU8();
            object = this.decodeExtension(size, 1);
          } else if (headByte === 200) {
            const size = this.lookU16();
            object = this.decodeExtension(size, 2);
          } else if (headByte === 201) {
            const size = this.lookU32();
            object = this.decodeExtension(size, 4);
          } else {
            throw new DecodeError_ts_1.DecodeError(`Unrecognized type byte: ${(0, prettyByte_ts_1.prettyByte)(headByte)}`);
          }
          this.complete();
          const stack = this.stack;
          while (stack.length > 0) {
            const state = stack.top();
            if (state.type === STATE_ARRAY) {
              state.array[state.position] = object;
              state.position++;
              if (state.position === state.size) {
                object = state.array;
                stack.release(state);
              } else {
                continue DECODE;
              }
            } else if (state.type === STATE_MAP_KEY) {
              if (object === "__proto__") {
                throw new DecodeError_ts_1.DecodeError("The key __proto__ is not allowed");
              }
              state.key = this.mapKeyConverter(object);
              state.type = STATE_MAP_VALUE;
              continue DECODE;
            } else {
              state.map[state.key] = object;
              state.readCount++;
              if (state.readCount === state.size) {
                object = state.map;
                stack.release(state);
              } else {
                state.key = null;
                state.type = STATE_MAP_KEY;
                continue DECODE;
              }
            }
          }
          return object;
        }
      }
      readHeadByte() {
        if (this.headByte === HEAD_BYTE_REQUIRED) {
          this.headByte = this.readU8();
        }
        return this.headByte;
      }
      complete() {
        this.headByte = HEAD_BYTE_REQUIRED;
      }
      readArraySize() {
        const headByte = this.readHeadByte();
        switch (headByte) {
          case 220:
            return this.readU16();
          case 221:
            return this.readU32();
          default: {
            if (headByte < 160) {
              return headByte - 144;
            } else {
              throw new DecodeError_ts_1.DecodeError(`Unrecognized array type byte: ${(0, prettyByte_ts_1.prettyByte)(headByte)}`);
            }
          }
        }
      }
      pushMapState(size) {
        if (size > this.maxMapLength) {
          throw new DecodeError_ts_1.DecodeError(`Max length exceeded: map length (${size}) > maxMapLengthLength (${this.maxMapLength})`);
        }
        this.stack.pushMapState(size);
      }
      pushArrayState(size) {
        if (size > this.maxArrayLength) {
          throw new DecodeError_ts_1.DecodeError(`Max length exceeded: array length (${size}) > maxArrayLength (${this.maxArrayLength})`);
        }
        this.stack.pushArrayState(size);
      }
      decodeString(byteLength, headerOffset) {
        if (!this.rawStrings || this.stateIsMapKey()) {
          return this.decodeUtf8String(byteLength, headerOffset);
        }
        return this.decodeBinary(byteLength, headerOffset);
      }
      /**
       * @throws {@link RangeError}
       */
      decodeUtf8String(byteLength, headerOffset) {
        if (byteLength > this.maxStrLength) {
          throw new DecodeError_ts_1.DecodeError(`Max length exceeded: UTF-8 byte length (${byteLength}) > maxStrLength (${this.maxStrLength})`);
        }
        if (this.bytes.byteLength < this.pos + headerOffset + byteLength) {
          throw MORE_DATA;
        }
        const offset = this.pos + headerOffset;
        let object;
        if (this.stateIsMapKey() && this.keyDecoder?.canBeCached(byteLength)) {
          object = this.keyDecoder.decode(this.bytes, offset, byteLength);
        } else {
          object = (0, utf8_ts_1.utf8Decode)(this.bytes, offset, byteLength);
        }
        this.pos += headerOffset + byteLength;
        return object;
      }
      stateIsMapKey() {
        if (this.stack.length > 0) {
          const state = this.stack.top();
          return state.type === STATE_MAP_KEY;
        }
        return false;
      }
      /**
       * @throws {@link RangeError}
       */
      decodeBinary(byteLength, headOffset) {
        if (byteLength > this.maxBinLength) {
          throw new DecodeError_ts_1.DecodeError(`Max length exceeded: bin length (${byteLength}) > maxBinLength (${this.maxBinLength})`);
        }
        if (!this.hasRemaining(byteLength + headOffset)) {
          throw MORE_DATA;
        }
        const offset = this.pos + headOffset;
        const object = this.bytes.subarray(offset, offset + byteLength);
        this.pos += headOffset + byteLength;
        return object;
      }
      decodeExtension(size, headOffset) {
        if (size > this.maxExtLength) {
          throw new DecodeError_ts_1.DecodeError(`Max length exceeded: ext length (${size}) > maxExtLength (${this.maxExtLength})`);
        }
        const extType = this.view.getInt8(this.pos + headOffset);
        const data = this.decodeBinary(
          size,
          headOffset + 1
          /* extType */
        );
        return this.extensionCodec.decode(data, extType, this.context);
      }
      lookU8() {
        return this.view.getUint8(this.pos);
      }
      lookU16() {
        return this.view.getUint16(this.pos);
      }
      lookU32() {
        return this.view.getUint32(this.pos);
      }
      readU8() {
        const value = this.view.getUint8(this.pos);
        this.pos++;
        return value;
      }
      readI8() {
        const value = this.view.getInt8(this.pos);
        this.pos++;
        return value;
      }
      readU16() {
        const value = this.view.getUint16(this.pos);
        this.pos += 2;
        return value;
      }
      readI16() {
        const value = this.view.getInt16(this.pos);
        this.pos += 2;
        return value;
      }
      readU32() {
        const value = this.view.getUint32(this.pos);
        this.pos += 4;
        return value;
      }
      readI32() {
        const value = this.view.getInt32(this.pos);
        this.pos += 4;
        return value;
      }
      readU64() {
        const value = (0, int_ts_1.getUint64)(this.view, this.pos);
        this.pos += 8;
        return value;
      }
      readI64() {
        const value = (0, int_ts_1.getInt64)(this.view, this.pos);
        this.pos += 8;
        return value;
      }
      readU64AsBigInt() {
        const value = this.view.getBigUint64(this.pos);
        this.pos += 8;
        return value;
      }
      readI64AsBigInt() {
        const value = this.view.getBigInt64(this.pos);
        this.pos += 8;
        return value;
      }
      readF32() {
        const value = this.view.getFloat32(this.pos);
        this.pos += 4;
        return value;
      }
      readF64() {
        const value = this.view.getFloat64(this.pos);
        this.pos += 8;
        return value;
      }
    };
    exports.Decoder = Decoder;
  }
});

// node_modules/@msgpack/msgpack/dist.cjs/decode.cjs
var require_decode = __commonJS({
  "node_modules/@msgpack/msgpack/dist.cjs/decode.cjs"(exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.decode = decode;
    exports.decodeMulti = decodeMulti;
    var Decoder_ts_1 = require_Decoder();
    function decode(buffer, options) {
      const decoder = new Decoder_ts_1.Decoder(options);
      return decoder.decode(buffer);
    }
    function decodeMulti(buffer, options) {
      const decoder = new Decoder_ts_1.Decoder(options);
      return decoder.decodeMulti(buffer);
    }
  }
});

// node_modules/@msgpack/msgpack/dist.cjs/utils/stream.cjs
var require_stream = __commonJS({
  "node_modules/@msgpack/msgpack/dist.cjs/utils/stream.cjs"(exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.isAsyncIterable = isAsyncIterable;
    exports.asyncIterableFromStream = asyncIterableFromStream;
    exports.ensureAsyncIterable = ensureAsyncIterable;
    function isAsyncIterable(object) {
      return object[Symbol.asyncIterator] != null;
    }
    async function* asyncIterableFromStream(stream) {
      const reader = stream.getReader();
      try {
        while (true) {
          const { done, value } = await reader.read();
          if (done) {
            return;
          }
          yield value;
        }
      } finally {
        reader.releaseLock();
      }
    }
    function ensureAsyncIterable(streamLike) {
      if (isAsyncIterable(streamLike)) {
        return streamLike;
      } else {
        return asyncIterableFromStream(streamLike);
      }
    }
  }
});

// node_modules/@msgpack/msgpack/dist.cjs/decodeAsync.cjs
var require_decodeAsync = __commonJS({
  "node_modules/@msgpack/msgpack/dist.cjs/decodeAsync.cjs"(exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.decodeAsync = decodeAsync;
    exports.decodeArrayStream = decodeArrayStream;
    exports.decodeMultiStream = decodeMultiStream;
    var Decoder_ts_1 = require_Decoder();
    var stream_ts_1 = require_stream();
    async function decodeAsync(streamLike, options) {
      const stream = (0, stream_ts_1.ensureAsyncIterable)(streamLike);
      const decoder = new Decoder_ts_1.Decoder(options);
      return decoder.decodeAsync(stream);
    }
    function decodeArrayStream(streamLike, options) {
      const stream = (0, stream_ts_1.ensureAsyncIterable)(streamLike);
      const decoder = new Decoder_ts_1.Decoder(options);
      return decoder.decodeArrayStream(stream);
    }
    function decodeMultiStream(streamLike, options) {
      const stream = (0, stream_ts_1.ensureAsyncIterable)(streamLike);
      const decoder = new Decoder_ts_1.Decoder(options);
      return decoder.decodeStream(stream);
    }
  }
});

// node_modules/@msgpack/msgpack/dist.cjs/index.cjs
var require_dist = __commonJS({
  "node_modules/@msgpack/msgpack/dist.cjs/index.cjs"(exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.decodeTimestampExtension = exports.encodeTimestampExtension = exports.decodeTimestampToTimeSpec = exports.encodeTimeSpecToTimestamp = exports.encodeDateToTimeSpec = exports.EXT_TIMESTAMP = exports.ExtData = exports.ExtensionCodec = exports.Encoder = exports.DecodeError = exports.Decoder = exports.decodeMultiStream = exports.decodeArrayStream = exports.decodeAsync = exports.decodeMulti = exports.decode = exports.encode = void 0;
    var encode_ts_1 = require_encode();
    Object.defineProperty(exports, "encode", { enumerable: true, get: function() {
      return encode_ts_1.encode;
    } });
    var decode_ts_1 = require_decode();
    Object.defineProperty(exports, "decode", { enumerable: true, get: function() {
      return decode_ts_1.decode;
    } });
    Object.defineProperty(exports, "decodeMulti", { enumerable: true, get: function() {
      return decode_ts_1.decodeMulti;
    } });
    var decodeAsync_ts_1 = require_decodeAsync();
    Object.defineProperty(exports, "decodeAsync", { enumerable: true, get: function() {
      return decodeAsync_ts_1.decodeAsync;
    } });
    Object.defineProperty(exports, "decodeArrayStream", { enumerable: true, get: function() {
      return decodeAsync_ts_1.decodeArrayStream;
    } });
    Object.defineProperty(exports, "decodeMultiStream", { enumerable: true, get: function() {
      return decodeAsync_ts_1.decodeMultiStream;
    } });
    var Decoder_ts_1 = require_Decoder();
    Object.defineProperty(exports, "Decoder", { enumerable: true, get: function() {
      return Decoder_ts_1.Decoder;
    } });
    var DecodeError_ts_1 = require_DecodeError();
    Object.defineProperty(exports, "DecodeError", { enumerable: true, get: function() {
      return DecodeError_ts_1.DecodeError;
    } });
    var Encoder_ts_1 = require_Encoder();
    Object.defineProperty(exports, "Encoder", { enumerable: true, get: function() {
      return Encoder_ts_1.Encoder;
    } });
    var ExtensionCodec_ts_1 = require_ExtensionCodec();
    Object.defineProperty(exports, "ExtensionCodec", { enumerable: true, get: function() {
      return ExtensionCodec_ts_1.ExtensionCodec;
    } });
    var ExtData_ts_1 = require_ExtData();
    Object.defineProperty(exports, "ExtData", { enumerable: true, get: function() {
      return ExtData_ts_1.ExtData;
    } });
    var timestamp_ts_1 = require_timestamp();
    Object.defineProperty(exports, "EXT_TIMESTAMP", { enumerable: true, get: function() {
      return timestamp_ts_1.EXT_TIMESTAMP;
    } });
    Object.defineProperty(exports, "encodeDateToTimeSpec", { enumerable: true, get: function() {
      return timestamp_ts_1.encodeDateToTimeSpec;
    } });
    Object.defineProperty(exports, "encodeTimeSpecToTimestamp", { enumerable: true, get: function() {
      return timestamp_ts_1.encodeTimeSpecToTimestamp;
    } });
    Object.defineProperty(exports, "decodeTimestampToTimeSpec", { enumerable: true, get: function() {
      return timestamp_ts_1.decodeTimestampToTimeSpec;
    } });
    Object.defineProperty(exports, "encodeTimestampExtension", { enumerable: true, get: function() {
      return timestamp_ts_1.encodeTimestampExtension;
    } });
    Object.defineProperty(exports, "decodeTimestampExtension", { enumerable: true, get: function() {
      return timestamp_ts_1.decodeTimestampExtension;
    } });
  }
});

// src/index.ts
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";

// src/server.ts
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";

// node_modules/@hotstuff-labs/ts-sdk/dist/utils/transports.util.js
var ENDPOINTS_URLS = {
  mainnet: {
    api: "https://testnet-api.hotstuff.trade/",
    rpc: "https://testnet-api.hotstuff.trade/",
    ws: "wss://testnet-api.hotstuff.trade/ws/"
  },
  testnet: {
    api: "https://testnet-api.hotstuff.trade/",
    rpc: "https://testnet-api.hotstuff.trade/",
    ws: "wss://testnet-api.hotstuff.trade/ws/"
  }
};
function mergeRequestInit(...inits) {
  const merged = inits.reduce((acc, init) => ({ ...acc, ...init }), {});
  const headersList = inits.map((init) => init.headers).filter((headers) => typeof headers === "object");
  if (headersList.length > 0) {
    merged.headers = mergeHeadersInit(...headersList);
  }
  const signals = inits.map((init) => init.signal).filter((signal) => signal instanceof AbortSignal);
  if (signals.length > 0) {
    merged.signal = signals.length > 1 ? AbortSignal.any(signals) : signals[0];
  }
  return merged;
}
function mergeHeadersInit(...inits) {
  if (inits.length === 0 || inits.length === 1) {
    return new Headers(inits[0]);
  }
  const merged = new Headers();
  for (const headers of inits) {
    if (Symbol.iterator in headers) {
      for (const [key, value] of headers) {
        merged.set(key, value);
      }
    } else {
      for (const key in headers) {
        if (headers.hasOwnProperty(key)) {
          merged.set(key, headers[key]);
        }
      }
    }
  }
  return merged;
}

// src/endpoints.ts
function normalizeBaseUrl(url) {
  const trimmed = url.trim();
  return trimmed.endsWith("/") ? trimmed : `${trimmed}/`;
}
function inferMainnetUrlFromTestnet(url) {
  return normalizeBaseUrl(url.replace("://testnet-", "://"));
}
function pickEndpoint(override, fallback) {
  if (override && override.trim().length > 0) {
    return normalizeBaseUrl(override);
  }
  return normalizeBaseUrl(fallback);
}
function getSdkDefaults() {
  const sdkTestnet = {
    api: normalizeBaseUrl(ENDPOINTS_URLS.testnet.api.toString()),
    rpc: normalizeBaseUrl(ENDPOINTS_URLS.testnet.rpc.toString()),
    ws: normalizeBaseUrl(ENDPOINTS_URLS.testnet.ws.toString())
  };
  const sdkMainnet = {
    api: normalizeBaseUrl(ENDPOINTS_URLS.mainnet.api.toString()),
    rpc: normalizeBaseUrl(ENDPOINTS_URLS.mainnet.rpc.toString()),
    ws: normalizeBaseUrl(ENDPOINTS_URLS.mainnet.ws.toString())
  };
  const sdkMainnetLooksBroken = sdkMainnet.api === sdkTestnet.api && sdkMainnet.rpc === sdkTestnet.rpc && sdkMainnet.ws === sdkTestnet.ws;
  const inferredMainnet = {
    api: inferMainnetUrlFromTestnet(sdkTestnet.api),
    rpc: inferMainnetUrlFromTestnet(sdkTestnet.rpc),
    ws: inferMainnetUrlFromTestnet(sdkTestnet.ws)
  };
  return {
    mainnet: sdkMainnetLooksBroken ? inferredMainnet : sdkMainnet,
    testnet: sdkTestnet
  };
}
function resolveEndpointConfig(env = process.env) {
  const defaults = getSdkDefaults();
  return {
    mainnet: {
      api: pickEndpoint(env.HOTSTUFF_MAINNET_API_URL, defaults.mainnet.api),
      rpc: pickEndpoint(env.HOTSTUFF_MAINNET_RPC_URL, defaults.mainnet.rpc),
      ws: pickEndpoint(env.HOTSTUFF_MAINNET_WS_URL, defaults.mainnet.ws)
    },
    testnet: {
      api: pickEndpoint(env.HOTSTUFF_TESTNET_API_URL, defaults.testnet.api),
      rpc: pickEndpoint(env.HOTSTUFF_TESTNET_RPC_URL, defaults.testnet.rpc),
      ws: pickEndpoint(env.HOTSTUFF_TESTNET_WS_URL, defaults.testnet.ws)
    }
  };
}
function getActiveEndpoints(config, isTestnet) {
  return isTestnet ? config.testnet : config.mainnet;
}

// src/config.ts
function envBool(v, fallback) {
  if (v === void 0) return fallback;
  return v.toLowerCase() === "true";
}
function envNum(v, fallback) {
  const n = Number(v);
  return Number.isFinite(n) && n > 0 ? n : fallback;
}
function loadConfig(env = process.env) {
  return {
    isTestnet: env.HOTSTUFF_TESTNET !== "false",
    accountAddress: env.HOTSTUFF_ACCOUNT_ADDRESS,
    privateKey: env.HOTSTUFF_PRIVATE_KEY,
    httpTimeoutMs: envNum(env.HOTSTUFF_HTTP_TIMEOUT_MS, 3e3),
    wsTimeoutMs: envNum(env.HOTSTUFF_WS_TIMEOUT_MS, 1e4),
    useWsReads: envBool(env.HOTSTUFF_USE_WS_READS, false),
    endpoints: resolveEndpointConfig(env)
  };
}

// node_modules/@hotstuff-labs/ts-sdk/dist/apis/exchange.api.js
import { privateKeyToAccount } from "viem/accounts";

// node_modules/@hotstuff-labs/ts-sdk/dist/utils/signing.util.js
var import_msgpack = __toESM(require_dist(), 1);
import { keccak256 } from "viem";
async function signAction(args, options) {
  const isTestnet = options?.isTestnet ?? false;
  const actionBytes = (0, import_msgpack.encode)(args.action);
  const payloadHash = keccak256(actionBytes);
  const domain = {
    name: "HotstuffCore",
    version: "1",
    chainId: 1,
    verifyingContract: "0x1234567890123456789012345678901234567890"
  };
  const types = {
    Action: [
      { name: "source", type: "string" },
      { name: "hash", type: "bytes32" },
      { name: "txType", type: "uint16" }
    ]
  };
  const eip712Message = {
    source: isTestnet ? "Testnet" : "Mainnet",
    hash: payloadHash,
    txType: args.txType
  };
  const signature = await args.wallet.signTypedData({
    domain,
    types,
    primaryType: "Action",
    message: eip712Message
  });
  return signature;
}

// node_modules/@hotstuff-labs/ts-sdk/dist/utils/nonce.util.js
var NonceManager = class {
  constructor() {
    this.lastNonce = 0;
  }
  getNonce() {
    let nonce = Date.now();
    if (nonce <= this.lastNonce) {
      nonce = ++this.lastNonce;
    } else {
      this.lastNonce = nonce;
    }
    return nonce;
  }
};

// node_modules/@hotstuff-labs/ts-sdk/dist/methods/exchange/op_codes.js
var EXCHANGE_OP_CODES = {
  /**---Account Actions---*/
  addAgent: 1201,
  revokeAgent: 1211,
  updatePerpInstrumentLeverage: 1203,
  approveBrokerFee: 1207,
  createReferralCode: 1208,
  setReferrer: 1209,
  claimReferralRewards: 1210,
  /**---Trading Actions---*/
  placeOrder: 1301,
  cancelByOid: 1302,
  cancelAll: 1311,
  cancelByCloid: 1312,
  /**---Collateral Actions---*/
  accountSpotWithdrawRequest: 1002,
  accountDerivativeWithdrawRequest: 1003,
  accountSpotBalanceTransferRequest: 1051,
  accountDerivativeBalanceTransferRequest: 1052,
  accountInternalBalanceTransferRequest: 1053,
  /**---Vault Actions---*/
  depositToVault: 1401,
  redeemFromVault: 1402
};

// node_modules/@hotstuff-labs/ts-sdk/dist/apis/exchange.api.js
var ExchangeClient = class {
  constructor(args) {
    this.transport = args.transport;
    this.wallet = args.wallet;
    this.nonce = args.nonce ?? new NonceManager().getNonce;
  }
  /**---Account Management Methods---*/
  async addAgent(params, execute = true, opts) {
    let nonce = await this.nonce();
    const agentAccount = privateKeyToAccount(params.agentPrivateKey);
    const agentSignature = await signAction({
      wallet: agentAccount,
      action: {
        signer: params.signer,
        nonce
      },
      txType: EXCHANGE_OP_CODES["addAgent"]
    });
    params = {
      agentName: params.agentName,
      agent: params.agent,
      forAccount: params.forAccount,
      signature: agentSignature,
      validUntil: params.validUntil,
      nonce
    };
    return this._executeAction({ action: "addAgent", params }, opts?.signal, execute);
  }
  async revokeAgent(params, opts) {
    return this._executeAction({ action: "revokeAgent", params }, opts?.signal);
  }
  async updatePerpInstrumentLeverage(params, opts) {
    return this._executeAction({ action: "updatePerpInstrumentLeverage", params }, opts?.signal);
  }
  async approveBrokerFee(params, opts) {
    return this._executeAction({ action: "approveBrokerFee", params }, opts?.signal);
  }
  async createReferralCode(params, opts) {
    return this._executeAction({ action: "createReferralCode", params }, opts?.signal);
  }
  async setReferrer(params, opts) {
    return this._executeAction({ action: "setReferrer", params }, opts?.signal);
  }
  async claimReferralRewards(params, opts) {
    return this._executeAction({ action: "claimReferralRewards", params }, opts?.signal);
  }
  /**---Trading Methods---*/
  async placeOrder(params, opts) {
    return this._executeAction({ action: "placeOrder", params }, opts?.signal);
  }
  async cancelByOid(params, opts) {
    return this._executeAction({ action: "cancelByOid", params }, opts?.signal);
  }
  async cancelByCloid(params, opts) {
    return this._executeAction({ action: "cancelByCloid", params }, opts?.signal);
  }
  async cancelAll(params, opts) {
    return this._executeAction({ action: "cancelAll", params }, opts?.signal);
  }
  /**---Collateral Transfer Methods---*/
  async accountSpotWithdrawRequest(params, opts) {
    return this._executeAction({ action: "accountSpotWithdrawRequest", params }, opts?.signal);
  }
  async accountDerivativeWithdrawRequest(params, opts) {
    return this._executeAction({ action: "accountDerivativeWithdrawRequest", params }, opts?.signal);
  }
  async accountSpotBalanceTransferRequest(params, opts) {
    return this._executeAction({ action: "accountSpotBalanceTransferRequest", params }, opts?.signal);
  }
  async accountDerivativeBalanceTransferRequest(params, opts) {
    return this._executeAction({ action: "accountDerivativeBalanceTransferRequest", params }, opts?.signal);
  }
  async accountInternalBalanceTransferRequest(params, opts) {
    return this._executeAction({ action: "accountInternalBalanceTransferRequest", params }, opts?.signal);
  }
  /**---Vault Methods---*/
  async depositToVault(params, opts) {
    return this._executeAction({ action: "depositToVault", params }, opts?.signal);
  }
  async redeemFromVault(params, opts) {
    return this._executeAction({ action: "redeemFromVault", params }, opts?.signal);
  }
  /**---Private Methods---*/
  async _executeAction(request, signal, execute = true) {
    const { action, params } = request;
    params.nonce = params.nonce ?? await this.nonce();
    let signature = await signAction({
      wallet: this.wallet,
      action: params,
      txType: EXCHANGE_OP_CODES[action]
    }, {
      isTestnet: this.transport.isTestnet
    });
    if (execute) {
      const response = await this.transport.request("exchange", {
        action: {
          data: params,
          type: String(EXCHANGE_OP_CODES[action])
        },
        signature,
        nonce: params.nonce
      }, signal);
      return response;
    }
    return { params, signature };
  }
};

// node_modules/@hotstuff-labs/ts-sdk/dist/apis/info.api.js
var InfoClient = class {
  constructor(args) {
    this.transport = args.transport;
  }
  /**---Global Info Endpoints---*/
  async oracle(params, signal) {
    const request = {
      method: "oracle",
      params
    };
    return this.transport.request("info", request, signal);
  }
  async supportedCollateral(params, signal) {
    const request = {
      method: "supportedCollateral",
      params
    };
    return this.transport.request("info", request, signal);
  }
  async instruments(params, signal) {
    const request = {
      method: "instruments",
      params
    };
    return this.transport.request("info", request, signal);
  }
  async ticker(params, signal) {
    const request = {
      method: "ticker",
      params
    };
    return this.transport.request("info", request, signal);
  }
  async orderbook(params, signal) {
    const request = {
      method: "orderbook",
      params
    };
    return this.transport.request("info", request, signal);
  }
  async trades(params, signal) {
    const request = {
      method: "trades",
      params
    };
    return this.transport.request("info", request, signal);
  }
  async mids(params, signal) {
    const request = {
      method: "mids",
      params
    };
    return this.transport.request("info", request, signal);
  }
  async bbo(params, signal) {
    const request = {
      method: "bbo",
      params
    };
    return this.transport.request("info", request, signal);
  }
  async chart(params, signal) {
    const request = {
      method: "chart",
      params
    };
    return this.transport.request("info", request, signal);
  }
  /**Account Info Endpoints */
  async openOrders(params, signal) {
    const request = {
      method: "openOrders",
      params
    };
    return this.transport.request("info", request, signal);
  }
  async positions(params, signal) {
    const request = {
      method: "positions",
      params
    };
    return this.transport.request("info", request, signal);
  }
  async accountSummary(params, signal) {
    const request = {
      method: "accountSummary",
      params
    };
    return this.transport.request("info", request, signal);
  }
  async referralSummary(params, signal) {
    const request = {
      method: "referralSummary",
      params
    };
    return this.transport.request("info", request, signal);
  }
  async userFeeInfo(params, signal) {
    const request = {
      method: "userFees",
      params
    };
    return this.transport.request("info", request, signal);
  }
  async accountHistory(params, signal) {
    const request = {
      method: "accountHistory",
      params
    };
    return this.transport.request("info", request, signal);
  }
  async orderHistory(params, signal) {
    const request = {
      method: "orderHistory",
      params
    };
    return this.transport.request("info", request, signal);
  }
  async tradeHistory(params, signal) {
    const request = {
      method: "fills",
      params
    };
    return this.transport.request("info", request, signal);
  }
  async fundingHistory(params, signal) {
    const request = {
      method: "fundingHistory",
      params
    };
    return this.transport.request("info", request, signal);
  }
  async transferHistory(params, signal) {
    const request = {
      method: "transferHistory",
      params
    };
    return this.transport.request("info", request, signal);
  }
  async instrumentLeverage(params, signal) {
    const request = {
      method: "instrumentLeverage",
      params
    };
    return this.transport.request("info", request, signal);
  }
  async agents(params, signal) {
    const request = {
      method: "allAgents",
      params
    };
    return this.transport.request("info", request, signal);
  }
  async accountInfo(params, signal) {
    const request = {
      method: "accountInfo",
      params
    };
    return this.transport.request("info", request, signal);
  }
  /**---Vault Info Endpoints---*/
  async vaults(params, signal) {
    const request = {
      method: "vaults",
      params
    };
    return this.transport.request("info", request, signal);
  }
  async subVaults(params, signal) {
    const request = {
      method: "subVaults",
      params
    };
    return this.transport.request("info", request, signal);
  }
  async vaultBalances(params, signal) {
    const request = {
      method: "vaultBalance",
      params
    };
    return this.transport.request("info", request, signal);
  }
  /** Explorer Info Endpoints */
  async blocks(params, signal) {
    const request = {
      method: "blocks",
      params
    };
    return this.transport.request("explorer", request, signal);
  }
  async blockDetails(params, signal) {
    const request = {
      method: "block",
      params
    };
    return this.transport.request("explorer", request, signal);
  }
  async transactions(params, signal) {
    const request = {
      method: "transactions",
      params
    };
    return this.transport.request("explorer", request, signal);
  }
  async transactionDetails(params, signal) {
    const request = {
      method: "transaction",
      params
    };
    return this.transport.request("explorer", request, signal);
  }
};

// node_modules/@hotstuff-labs/ts-sdk/dist/apis/subscription.api.js
var SubscriptionClient = class {
  constructor(args) {
    this.transport = args.transport;
  }
  /** Global Subscription Endpoints */
  async ticker(params, listener) {
    const result = await this.transport.subscribe("ticker", params, listener);
    const resultWithId = result;
    return {
      subscriptionId: resultWithId.subscriptionId,
      unsubscribe: () => this.transport.unsubscribe(resultWithId.subscriptionId)
    };
  }
  async mids(params, listener) {
    const result = await this.transport.subscribe("mids", params, listener);
    const resultWithId = result;
    return {
      subscriptionId: resultWithId.subscriptionId,
      unsubscribe: () => this.transport.unsubscribe(resultWithId.subscriptionId)
    };
  }
  async bbo(params, listener) {
    const result = await this.transport.subscribe("bbo", params, listener);
    const resultWithId = result;
    return {
      subscriptionId: resultWithId.subscriptionId,
      unsubscribe: () => this.transport.unsubscribe(resultWithId.subscriptionId)
    };
  }
  async orderbook(params, listener) {
    const request = {
      instrumentId: params.instrumentId,
      symbol: params.instrumentId
    };
    const result = await this.transport.subscribe("orderbook", request, listener);
    const resultWithId = result;
    return {
      subscriptionId: resultWithId.subscriptionId,
      unsubscribe: () => this.transport.unsubscribe(resultWithId.subscriptionId)
    };
  }
  async trade(params, listener) {
    const request = {
      instrumentId: params.instrumentId,
      symbol: params.instrumentId
    };
    const result = await this.transport.subscribe("trade", request, listener);
    const resultWithId = result;
    return {
      subscriptionId: resultWithId.subscriptionId,
      unsubscribe: () => this.transport.unsubscribe(resultWithId.subscriptionId)
    };
  }
  async index(listener) {
    const result = await this.transport.subscribe("index", {}, listener);
    const resultWithId = result;
    return {
      subscriptionId: resultWithId.subscriptionId,
      unsubscribe: () => this.transport.unsubscribe(resultWithId.subscriptionId)
    };
  }
  async chart(params, listener) {
    const result = await this.transport.subscribe("chart", params, listener);
    const resultWithId = result;
    return {
      subscriptionId: resultWithId.subscriptionId,
      unsubscribe: () => this.transport.unsubscribe(resultWithId.subscriptionId)
    };
  }
  /** Account Subscription Endpoints */
  async accountOrderUpdates(params, listener) {
    const request = {
      address: params.address,
      user: params.address
    };
    const result = await this.transport.subscribe("order", request, listener);
    const resultWithId = result;
    return {
      subscriptionId: resultWithId.subscriptionId,
      unsubscribe: () => this.transport.unsubscribe(resultWithId.subscriptionId)
    };
  }
  async accountBalanceUpdates(params, listener) {
    const request = {
      address: params.address,
      user: params.address
    };
    const result = await this.transport.subscribe("balance", request, listener);
    const resultWithId = result;
    return {
      subscriptionId: resultWithId.subscriptionId,
      unsubscribe: () => this.transport.unsubscribe(resultWithId.subscriptionId)
    };
  }
  async positions(params, listener) {
    const request = {
      address: params.address,
      user: params.address
    };
    const result = await this.transport.subscribe("position", request, listener);
    const resultWithId = result;
    return {
      subscriptionId: resultWithId.subscriptionId,
      unsubscribe: () => this.transport.unsubscribe(resultWithId.subscriptionId)
    };
  }
  async fills(params, listener) {
    const request = {
      address: params.address,
      user: params.address
    };
    const result = await this.transport.subscribe("fills", request, listener);
    const resultWithId = result;
    return {
      subscriptionId: resultWithId.subscriptionId,
      unsubscribe: () => this.transport.unsubscribe(resultWithId.subscriptionId)
    };
  }
  async accountSummary(params, listener) {
    const request = {
      user: params.user
    };
    const result = await this.transport.subscribe("account_summary", request, listener);
    const resultWithId = result;
    return {
      subscriptionId: resultWithId.subscriptionId,
      unsubscribe: () => this.transport.unsubscribe(resultWithId.subscriptionId)
    };
  }
  /** Explorer Subscription Endpoints */
  async blocks(params, listener) {
    const request = {};
    const result = await this.transport.subscribe("blocks", request, listener);
    const resultWithId = result;
    return {
      subscriptionId: resultWithId.subscriptionId,
      unsubscribe: () => this.transport.unsubscribe(resultWithId.subscriptionId)
    };
  }
  async transactions(params, listener) {
    const request = {};
    const result = await this.transport.subscribe("transactions", request, listener);
    const resultWithId = result;
    return {
      subscriptionId: resultWithId.subscriptionId,
      unsubscribe: () => this.transport.unsubscribe(resultWithId.subscriptionId)
    };
  }
};

// node_modules/@hotstuff-labs/ts-sdk/dist/transports/http.transport.js
var HttpTransport = class {
  constructor(options) {
    this.isTestnet = options?.isTestnet ?? false;
    this.timeout = options?.timeout ?? 3e3;
    this.server = {
      mainnet: {
        api: options?.server?.mainnet?.api ?? ENDPOINTS_URLS.mainnet.api,
        rpc: options?.server?.mainnet?.rpc ?? ENDPOINTS_URLS.mainnet.rpc
      },
      testnet: {
        api: options?.server?.testnet?.api ?? ENDPOINTS_URLS.testnet.api,
        rpc: options?.server?.testnet?.rpc ?? ENDPOINTS_URLS.testnet.rpc
      }
    };
    this.fetchOptions = options?.fetchOptions ?? {};
    this.onRequest = options?.onRequest;
    this.onResponse = options?.onResponse;
  }
  async request(endpoint, payload, signal, method = "POST") {
    try {
      const url = new URL(endpoint, this.server[this.isTestnet ? "testnet" : "mainnet"][endpoint === "explorer" ? "rpc" : "api"]);
      const init = mergeRequestInit({
        body: method === "POST" ? JSON.stringify(payload) : void 0,
        headers: {
          "Accept-Encoding": "gzip, deflate, br, zstd",
          "Content-Type": "application/json"
        },
        keepalive: true,
        method,
        signal: this.timeout ? AbortSignal.timeout(this.timeout) : void 0
      }, this.fetchOptions, { signal });
      let request = new Request(url, init);
      if (this.onRequest) {
        const customRequest = await this.onRequest(request);
        if (customRequest instanceof Request)
          request = customRequest;
      }
      let response = await fetch(request);
      if (this.onResponse) {
        const customResponse = await this.onResponse(response);
        if (customResponse instanceof Response)
          response = customResponse;
      }
      if (!response.ok || !response.headers.get("Content-Type")?.includes("application/json")) {
        const body2 = await response.text().catch(() => void 0);
        throw new Error(body2);
      }
      const body = await response.json();
      if (body?.type === "error") {
        throw new Error(body?.message);
      }
      return body;
    } catch (error) {
      if (error instanceof Error)
        throw error;
      throw new Error(error);
    }
  }
};

// node_modules/@hotstuff-labs/ts-sdk/dist/types/transports/index.js
var transports_exports = {};
__export(transports_exports, {
  WSEventType: () => WSEventType,
  WSMethod: () => WSMethod
});

// node_modules/@hotstuff-labs/ts-sdk/dist/types/transports/websocket.type.js
var WSMethod;
(function(WSMethod2) {
  WSMethod2["SUBSCRIBE"] = "subscribe";
  WSMethod2["UNSUBSCRIBE"] = "unsubscribe";
  WSMethod2["PING"] = "ping";
})(WSMethod || (WSMethod = {}));
var WSEventType;
(function(WSEventType2) {
  WSEventType2["DEPOSIT"] = "deposit";
  WSEventType2["SPOT_WITHDRAW"] = "spot_withdraw";
  WSEventType2["DERIVATIVE_WITHDRAW"] = "derivative_withdraw";
  WSEventType2["ORDER_UPDATE"] = "order_update";
  WSEventType2["TRADE"] = "trade";
  WSEventType2["SPOT_TRADE"] = "spot_trade";
  WSEventType2["ACCOUNT_BALANCE_UPDATED"] = "account_balance_updated";
  WSEventType2["ACCOUNT_SPOT_BALANCE_UPDATED"] = "account_spot_balance_updated";
  WSEventType2["AGENT_REGISTERED"] = "agent_registered";
  WSEventType2["ACCOUNT_UPDATED"] = "account_updated";
  WSEventType2["PRICE_INDEX_ADDED"] = "price_index_added";
  WSEventType2["ORACLE_FEED_UPDATED"] = "oracle_feed_updated";
  WSEventType2["PERP_LISTED"] = "perp_listed";
  WSEventType2["SPOT_LISTED"] = "spot_listed";
  WSEventType2["OPTION_LISTED"] = "option_listed";
  WSEventType2["SPOT_BALANCE_TRANSFER"] = "spot_balance_transfer";
  WSEventType2["DERIVATIVE_BALANCE_TRANSFER"] = "derivative_balance_transfer";
  WSEventType2["INTERNAL_BALANCE_TRANSFER"] = "internal_balance_transfer";
  WSEventType2["SPOT_SWAP_TRIGGERED"] = "spot_swap_triggered";
  WSEventType2["EVM_COLLATERAL_ADDED"] = "evm_collateral_added";
  WSEventType2["INDEXER_BOOK_UPDATE"] = "indexer_book_update";
  WSEventType2["TICKER"] = "ticker";
})(WSEventType || (WSEventType = {}));

// node_modules/@hotstuff-labs/ts-sdk/dist/transports/websocket.transport.js
var WebSocketTransport = class {
  constructor(options) {
    this.ws = null;
    this.reconnectAttempts = 0;
    this.maxReconnectAttempts = 5;
    this.reconnectDelay = 1e3;
    this.messageQueue = [];
    this.messageIdCounter = 0;
    this.subscriptions = /* @__PURE__ */ new Map();
    this.subscriptionCallbacks = /* @__PURE__ */ new Map();
    this.connectionPromise = null;
    this.isTestnet = options?.isTestnet ?? false;
    this.timeout = options?.timeout ?? 1e4;
    this.server = {
      mainnet: options?.server?.mainnet ?? ENDPOINTS_URLS.mainnet.ws,
      testnet: options?.server?.testnet ?? ENDPOINTS_URLS.testnet.ws
    };
    this.keepAlive = options?.keepAlive ?? {
      interval: 3e4,
      timeout: 1e4
    };
    const autoConnect = options?.autoConnect ?? true;
    if (autoConnect) {
      this.connectionPromise = this.connect().catch((error) => {
        console.error("Auto-connect failed:", error);
      });
    }
  }
  async ensureConnected() {
    if (this.ws?.readyState === WebSocket.OPEN) {
      return;
    }
    if (this.connectionPromise) {
      await this.connectionPromise;
      return;
    }
    this.connectionPromise = this.connect();
    await this.connectionPromise;
  }
  cleanup() {
    if (this.keepAliveInterval) {
      clearInterval(this.keepAliveInterval);
      this.keepAliveInterval = void 0;
    }
    this.messageQueue.forEach(({ reject }) => {
      reject(new Error("WebSocket disconnected"));
    });
    this.messageQueue = [];
  }
  startKeepAlive() {
    if (this.keepAlive?.interval) {
      this.keepAliveInterval = setInterval(() => {
        this.ping();
      }, this.keepAlive.interval);
    }
  }
  handleIncomingMessage(message) {
    if (message.id && (message.result !== void 0 || message.error !== void 0)) {
      this.handleJSONRPCResponse(message);
      return;
    }
    if (message.method && message.params && !message.id) {
      this.handleJSONRPCNotification(message);
      return;
    }
  }
  handleJSONRPCResponse(response) {
    const queueItem = this.messageQueue.find((item) => item.id === response.id);
    if (queueItem) {
      this.messageQueue = this.messageQueue.filter((item) => item.id !== response.id);
      if (response.error) {
        queueItem.reject(new Error(`JSON-RPC Error ${response.error.code}: ${response.error.message}`));
      } else {
        queueItem.resolve(response.result);
      }
    } else {
    }
  }
  handleJSONRPCNotification(notification) {
    if ((notification.method === "subscription" || notification.method === "event") && notification.params) {
      const { channel, data } = notification.params;
      let notificationSent = false;
      this.subscriptions.forEach((subscription, subscriptionId) => {
        if (subscription.channel === channel) {
          const callback = this.subscriptionCallbacks.get(subscriptionId);
          if (callback) {
            callback({
              channel,
              data,
              timestamp: Date.now()
            });
            notificationSent = true;
          } else {
            console.warn(`No callback found for subscription ${subscriptionId}`);
          }
        }
      });
    } else {
      console.warn(`Unknown notification method: ${notification.method}`);
    }
  }
  async sendJSONRPCMessage(message) {
    await this.ensureConnected();
    return new Promise((resolve, reject) => {
      if (message.id === null || message.id === void 0) {
        message.id = (++this.messageIdCounter).toString();
      }
      if (message.id !== void 0) {
        this.messageQueue.push({
          id: message.id,
          resolve,
          reject
        });
        if (this.timeout) {
          setTimeout(() => {
            this.messageQueue = this.messageQueue.filter((qm) => qm.id !== message.id);
            reject(new Error("Request timeout"));
          }, this.timeout);
        }
      }
      const messageStr = JSON.stringify(message);
      this.ws.send(messageStr);
      if (message.id === void 0) {
        resolve({});
      }
    });
  }
  formatSubscriptionParams(channel, payload) {
    const payloadObj = payload;
    const subscription = {
      channel,
      ...payloadObj
    };
    return subscription;
  }
  async subscribeToChannels(params) {
    const message = {
      jsonrpc: "2.0",
      method: transports_exports.WSMethod.SUBSCRIBE,
      params,
      id: (++this.messageIdCounter).toString()
    };
    return this.sendJSONRPCMessage(message);
  }
  async unsubscribeFromChannels(channels) {
    const message = {
      jsonrpc: "2.0",
      method: transports_exports.WSMethod.UNSUBSCRIBE,
      params: channels,
      // Send channels directly as array, not wrapped in object
      id: (++this.messageIdCounter).toString()
    };
    return this.sendJSONRPCMessage(message);
  }
  isConnected() {
    return this.ws?.readyState === WebSocket.OPEN;
  }
  async connect() {
    return new Promise((resolve, reject) => {
      try {
        const url = this.isTestnet ? this.server.testnet : this.server.mainnet;
        this.ws = new WebSocket(url);
        this.ws.onopen = () => {
          this.reconnectAttempts = 0;
          this.startKeepAlive();
          resolve();
        };
        this.ws.onmessage = (event) => {
          try {
            const message = JSON.parse(event.data);
            this.handleIncomingMessage(message);
          } catch (error) {
            console.error("Failed to parse WebSocket message:", error);
          }
        };
        this.ws.onclose = (event) => {
          this.cleanup();
          if (this.reconnectAttempts < this.maxReconnectAttempts) {
            setTimeout(() => {
              this.reconnectAttempts++;
              this.connect();
            }, this.reconnectDelay * this.reconnectAttempts);
          }
        };
        this.ws.onerror = (error) => {
          console.error("WebSocket error:", error);
          reject(error);
        };
      } catch (error) {
        reject(error);
      }
    });
  }
  async disconnect() {
    this.cleanup();
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
  }
  async ping() {
    await this.ensureConnected();
    const message = {
      jsonrpc: "2.0",
      method: transports_exports.WSMethod.PING,
      id: (++this.messageIdCounter).toString()
    };
    return this.sendJSONRPCMessage(message);
  }
  async request(type, payload, signal) {
    throw new Error("Use specific WebSocket methods (subscribe, unsubscribe, ping) instead of generic request");
  }
  async subscribe(channel, payload, listener) {
    await this.ensureConnected();
    const subscriptionId = `${channel}_${Date.now()}`;
    const subscription = {
      id: subscriptionId,
      channel,
      symbol: payload?.instrumentId || payload?.symbol,
      params: payload,
      timestamp: Date.now()
    };
    this.subscriptionCallbacks.set(subscriptionId, (data) => {
      const customEvent = new CustomEvent("subscription", {
        detail: data.data
      });
      listener(customEvent);
    });
    try {
      const subscriptionParams = this.formatSubscriptionParams(channel, payload);
      const result = await this.subscribeToChannels(subscriptionParams);
      if (result.status === "subscribed" && result.channels && result.channels.length > 0) {
        const serverChannel = result.channels[0];
        subscription.channel = serverChannel;
        this.subscriptions.set(subscriptionId, subscription);
        return { ...result, subscriptionId };
      } else if (result.status === "failed" || result.status === "error") {
        this.subscriptionCallbacks.delete(subscriptionId);
        const errorMsg = result.error || `Subscription ${result.status}`;
        throw new Error(`Server rejected subscription: ${errorMsg}`);
      } else if (!result.channels || result.channels.length === 0) {
        this.subscriptionCallbacks.delete(subscriptionId);
        throw new Error(`Server returned no channels for subscription`);
      } else {
        this.subscriptionCallbacks.delete(subscriptionId);
        throw new Error(`Unknown subscription status: ${result.status}`);
      }
    } catch (error) {
      this.subscriptions.delete(subscriptionId);
      this.subscriptionCallbacks.delete(subscriptionId);
      console.error(`Subscription failed for ${channel}:`, error);
      throw error;
    }
  }
  async unsubscribe(subscriptionId) {
    const subscription = this.subscriptions.get(subscriptionId);
    if (!subscription) {
      throw new Error(`Subscription ${subscriptionId} not found`);
    }
    try {
      if (this.isConnected()) {
        await this.unsubscribeFromChannels([subscription.channel]);
      }
      this.subscriptions.delete(subscriptionId);
      this.subscriptionCallbacks.delete(subscriptionId);
    } catch (error) {
      console.error(`Failed to unsubscribe from ${subscription.channel}:`, error);
      this.subscriptions.delete(subscriptionId);
      this.subscriptionCallbacks.delete(subscriptionId);
      throw error;
    }
  }
  getSubscriptions() {
    return Array.from(this.subscriptions.values());
  }
};

// src/context.ts
import { privateKeyToAccount as privateKeyToAccount2 } from "viem/accounts";
function createHotstuffContext(config) {
  const httpTransport = new HttpTransport({
    isTestnet: config.isTestnet,
    timeout: config.httpTimeoutMs,
    server: {
      mainnet: {
        api: config.endpoints.mainnet.api,
        rpc: config.endpoints.mainnet.rpc
      },
      testnet: {
        api: config.endpoints.testnet.api,
        rpc: config.endpoints.testnet.rpc
      }
    }
  });
  const wsTransport = new WebSocketTransport({
    isTestnet: config.isTestnet,
    timeout: config.wsTimeoutMs,
    autoConnect: true,
    server: {
      mainnet: config.endpoints.mainnet.ws,
      testnet: config.endpoints.testnet.ws
    }
  });
  const infoTransport = config.useWsReads ? wsTransport : httpTransport;
  const info = new InfoClient({ transport: infoTransport });
  const subscription = new SubscriptionClient({ transport: wsTransport });
  let exchange;
  if (config.privateKey) {
    const wallet = privateKeyToAccount2(config.privateKey);
    exchange = new ExchangeClient({ transport: httpTransport, wallet });
  }
  return {
    config,
    httpTransport,
    wsTransport,
    info,
    subscription,
    exchange
  };
}

// src/instructions.ts
var SERVER_INSTRUCTIONS = `
You are connected to the Hotstuff trading MCP server.

Safety:
- Read-only tools may be used without asking the user each time.
- Every write tool (orders, cancels, leverage, agents, broker approval, referral actions, withdrawals, balance transfers, vault deposit/redeem) requires explicit human confirmation before you call it. Summarize side, size, price, instrument, expiry, and addresses first.
- High-risk actions: cancel_all_orders, withdrawals (request_spot_withdraw, request_derivative_withdraw), external transfers (transfer_spot_balance, transfer_derivative_balance), vault deposit_to_vault and redeem_from_vault \u2014 require extra confirmation.
- If HOTSTUFF_PRIVATE_KEY is not configured, write tools are unavailable; explain that the user must set it in the MCP server environment.

Data:
- For account-scoped reads, omit user to use HOTSTUFF_ACCOUNT_ADDRESS when set.
`.trim();

// src/resources/endpoints.ts
function getEndpointsPayload(isTestnet, endpoints) {
  const urls = getActiveEndpoints(endpoints, isTestnet);
  return {
    network: isTestnet ? "testnet" : "mainnet",
    api: urls.api,
    rpc: urls.rpc,
    ws: urls.ws
  };
}

// src/resources/endpointsResource.ts
function registerEndpointsResource(server2, ctx) {
  server2.registerResource(
    "endpoints",
    "hotstuff://endpoints",
    {
      title: "API endpoints",
      description: "Default HTTP and WebSocket bases for the configured network.",
      mimeType: "application/json"
    },
    async () => ({
      contents: [
        {
          uri: "hotstuff://endpoints",
          mimeType: "application/json",
          text: JSON.stringify(
            {
              fetchedAt: Date.now(),
              isTestnet: ctx.config.isTestnet,
              ...getEndpointsPayload(ctx.config.isTestnet, ctx.config.endpoints)
            },
            null,
            2
          )
        }
      ]
    })
  );
}

// src/constants/exchangeOpCodes.ts
var EXCHANGE_OP_CODES2 = {
  addAgent: 1201,
  revokeAgent: 1211,
  updatePerpInstrumentLeverage: 1203,
  approveBrokerFee: 1207,
  createReferralCode: 1208,
  setReferrer: 1209,
  claimReferralRewards: 1210,
  placeOrder: 1301,
  cancelByOid: 1302,
  cancelAll: 1311,
  cancelByCloid: 1312,
  accountSpotWithdrawRequest: 1002,
  accountDerivativeWithdrawRequest: 1003,
  accountSpotBalanceTransferRequest: 1051,
  accountDerivativeBalanceTransferRequest: 1052,
  accountInternalBalanceTransferRequest: 1053,
  depositToVault: 1401,
  redeemFromVault: 1402
};

// src/resources/exchangeOpCodes.ts
function registerExchangeOpCodesResource(server2) {
  server2.registerResource(
    "exchange_opcodes",
    "hotstuff://exchange-opcodes",
    {
      title: "Exchange op codes",
      description: "Numeric op codes for signed exchange actions.",
      mimeType: "application/json"
    },
    async () => ({
      contents: [
        {
          uri: "hotstuff://exchange-opcodes",
          mimeType: "application/json",
          text: JSON.stringify(EXCHANGE_OP_CODES2, null, 2)
        }
      ]
    })
  );
}

// src/resources/instruments.ts
function registerInstrumentsResource(server2, ctx) {
  server2.registerResource(
    "instruments",
    "hotstuff://instruments",
    {
      title: "Instruments snapshot",
      description: "Live instruments list (type=all).",
      mimeType: "application/json"
    },
    async () => {
      const data = await ctx.info.instruments({ type: "all" });
      return {
        contents: [
          {
            uri: "hotstuff://instruments",
            mimeType: "application/json",
            text: JSON.stringify(
              {
                fetchedAt: Date.now(),
                isTestnet: ctx.config.isTestnet,
                data
              },
              null,
              2
            )
          }
        ]
      };
    }
  );
}

// src/resources/supportedCollateral.ts
function registerSupportedCollateralResource(server2, ctx) {
  server2.registerResource(
    "supported_collateral",
    "hotstuff://supported-collateral",
    {
      title: "Supported collateral",
      description: "Live supported collateral snapshot.",
      mimeType: "application/json"
    },
    async () => {
      const data = await ctx.info.supportedCollateral({});
      return {
        contents: [
          {
            uri: "hotstuff://supported-collateral",
            mimeType: "application/json",
            text: JSON.stringify(
              {
                fetchedAt: Date.now(),
                isTestnet: ctx.config.isTestnet,
                data
              },
              null,
              2
            )
          }
        ]
      };
    }
  );
}

// src/resources/index.ts
function registerAllResources(server2, ctx) {
  registerExchangeOpCodesResource(server2);
  registerEndpointsResource(server2, ctx);
  registerInstrumentsResource(server2, ctx);
  registerSupportedCollateralResource(server2, ctx);
}

// src/tools/account/getAccountHistory.ts
import { z } from "zod";

// src/utils/redact.ts
function redactErrorMessage(message) {
  return message.replace(/\b0x[a-fA-F0-9]{64,}\b/g, "0x[REDACTED]").replace(/\b[a-fA-F0-9]{80,}\b/g, "[REDACTED]");
}

// src/utils/result.ts
function ok(data) {
  return { content: [{ type: "text", text: JSON.stringify(data, null, 2) }] };
}
function toolError(action, error) {
  const raw = error instanceof Error ? error.message : String(error);
  const msg = redactErrorMessage(raw);
  return {
    isError: true,
    content: [{ type: "text", text: `[${action} error] ${msg}` }]
  };
}

// src/utils/user.ts
function resolveUser(ctx, user) {
  const addr = user ?? ctx.config.accountAddress;
  if (!addr) {
    throw new Error("Provide `user` or set HOTSTUFF_ACCOUNT_ADDRESS.");
  }
  return addr;
}
function expiresIn(seconds) {
  return Math.floor(Date.now() / 1e3) + seconds;
}

// src/tools/annotations.ts
var READ_ONLY_ANNOTATION = { readOnlyHint: true };
var DESTRUCTIVE_ANNOTATION = { destructiveHint: true };

// src/tools/account/getAccountHistory.ts
function registerGetAccountHistoryTool(server2, ctx) {
  server2.registerTool(
    "get_account_history",
    {
      title: "Get account history",
      description: "Historical PnL / volume / equity snapshots. Omit user to use HOTSTUFF_ACCOUNT_ADDRESS.",
      inputSchema: z.object({
        user: z.string().optional()
      }),
      annotations: READ_ONLY_ANNOTATION
    },
    async ({ user }) => {
      try {
        return ok(await ctx.info.accountHistory({ user: resolveUser(ctx, user) }));
      } catch (e) {
        return toolError("get_account_history", e);
      }
    }
  );
}

// src/tools/account/getAccountInfo.ts
import { z as z2 } from "zod";
function registerGetAccountInfoTool(server2, ctx) {
  server2.registerTool(
    "get_account_info",
    {
      title: "Get account info",
      description: "Account metadata and referral info. Omit user to use HOTSTUFF_ACCOUNT_ADDRESS.",
      inputSchema: z2.object({
        user: z2.string().optional().describe("Optional 0x address."),
        collateralID: z2.number().int().min(0).optional(),
        includeHistory: z2.boolean().optional()
      }),
      annotations: READ_ONLY_ANNOTATION
    },
    async ({ user, collateralID, includeHistory }) => {
      try {
        return ok(
          await ctx.info.accountInfo({
            user: resolveUser(ctx, user),
            collateralID,
            includeHistory
          })
        );
      } catch (e) {
        return toolError("get_account_info", e);
      }
    }
  );
}

// src/tools/account/getAccountSummary.ts
import { z as z3 } from "zod";
function registerGetAccountSummaryTool(server2, ctx) {
  server2.registerTool(
    "get_account_summary",
    {
      title: "Get account summary",
      description: "Full account summary (balances, positions, margin). Omit user to use HOTSTUFF_ACCOUNT_ADDRESS.",
      inputSchema: z3.object({
        user: z3.string().optional().describe("Optional 0x address.")
      }),
      annotations: READ_ONLY_ANNOTATION
    },
    async ({ user }) => {
      try {
        return ok(await ctx.info.accountSummary({ user: resolveUser(ctx, user) }));
      } catch (e) {
        return toolError("get_account_summary", e);
      }
    }
  );
}

// src/tools/account/getAgents.ts
import { z as z4 } from "zod";
function registerGetAgentsTool(server2, ctx) {
  server2.registerTool(
    "get_agents",
    {
      title: "Get agents",
      description: "Registered trading agents. Omit user to use HOTSTUFF_ACCOUNT_ADDRESS.",
      inputSchema: z4.object({
        user: z4.string().optional()
      }),
      annotations: READ_ONLY_ANNOTATION
    },
    async ({ user }) => {
      try {
        return ok(await ctx.info.agents({ user: resolveUser(ctx, user) }));
      } catch (e) {
        return toolError("get_agents", e);
      }
    }
  );
}

// src/tools/account/getFundingHistory.ts
import { z as z5 } from "zod";
function registerGetFundingHistoryTool(server2, ctx) {
  server2.registerTool(
    "get_funding_history",
    {
      title: "Get funding history",
      description: "Funding payments. Omit user to use HOTSTUFF_ACCOUNT_ADDRESS.",
      inputSchema: z5.object({
        user: z5.string().optional()
      }),
      annotations: READ_ONLY_ANNOTATION
    },
    async ({ user }) => {
      try {
        return ok(await ctx.info.fundingHistory({ user: resolveUser(ctx, user) }));
      } catch (e) {
        return toolError("get_funding_history", e);
      }
    }
  );
}

// src/tools/account/getInstrumentLeverage.ts
import { z as z6 } from "zod";
function registerGetInstrumentLeverageTool(server2, ctx) {
  server2.registerTool(
    "get_instrument_leverage",
    {
      title: "Get instrument leverage",
      description: "Leverage for one instrument. Omit user to use HOTSTUFF_ACCOUNT_ADDRESS.",
      inputSchema: z6.object({
        user: z6.string().optional(),
        symbol: z6.string().min(1).describe("e.g. BTC-PERP.")
      }),
      annotations: READ_ONLY_ANNOTATION
    },
    async ({ user, symbol }) => {
      try {
        return ok(await ctx.info.instrumentLeverage({ user: resolveUser(ctx, user), symbol }));
      } catch (e) {
        return toolError("get_instrument_leverage", e);
      }
    }
  );
}

// src/tools/account/getOpenOrders.ts
import { z as z7 } from "zod";
function registerGetOpenOrdersTool(server2, ctx) {
  server2.registerTool(
    "get_open_orders",
    {
      title: "Get open orders",
      description: "Open orders. Omit user to use HOTSTUFF_ACCOUNT_ADDRESS.",
      inputSchema: z7.object({
        user: z7.string().optional(),
        page: z7.number().int().positive().optional(),
        limit: z7.number().int().positive().optional()
      }),
      annotations: READ_ONLY_ANNOTATION
    },
    async ({ user, page, limit }) => {
      try {
        return ok(await ctx.info.openOrders({ user: resolveUser(ctx, user), page, limit }));
      } catch (e) {
        return toolError("get_open_orders", e);
      }
    }
  );
}

// src/tools/account/getOrderHistory.ts
import { z as z8 } from "zod";
function registerGetOrderHistoryTool(server2, ctx) {
  server2.registerTool(
    "get_order_history",
    {
      title: "Get order history",
      description: "Historical orders. Omit user to use HOTSTUFF_ACCOUNT_ADDRESS.",
      inputSchema: z8.object({
        user: z8.string().optional(),
        instrumentId: z8.string().optional(),
        limit: z8.number().int().positive().optional()
      }),
      annotations: READ_ONLY_ANNOTATION
    },
    async ({ user, instrumentId, limit }) => {
      try {
        return ok(
          await ctx.info.orderHistory({
            user: resolveUser(ctx, user),
            instrumentId,
            limit
          })
        );
      } catch (e) {
        return toolError("get_order_history", e);
      }
    }
  );
}

// src/tools/account/getPositions.ts
import { z as z9 } from "zod";
function registerGetPositionsTool(server2, ctx) {
  server2.registerTool(
    "get_positions",
    {
      title: "Get positions",
      description: "Open positions. Omit user to use HOTSTUFF_ACCOUNT_ADDRESS.",
      inputSchema: z9.object({
        user: z9.string().optional(),
        instrument: z9.string().optional().describe("Optional symbol filter, e.g. BTC-PERP.")
      }),
      annotations: READ_ONLY_ANNOTATION
    },
    async ({ user, instrument }) => {
      try {
        return ok(await ctx.info.positions({ user: resolveUser(ctx, user), instrument }));
      } catch (e) {
        return toolError("get_positions", e);
      }
    }
  );
}

// src/tools/account/getReferralSummary.ts
import { z as z10 } from "zod";
function registerGetReferralSummaryTool(server2, ctx) {
  server2.registerTool(
    "get_referral_summary",
    {
      title: "Get referral summary",
      description: "Referral codes and rewards. Omit user to use HOTSTUFF_ACCOUNT_ADDRESS.",
      inputSchema: z10.object({
        user: z10.string().optional()
      }),
      annotations: READ_ONLY_ANNOTATION
    },
    async ({ user }) => {
      try {
        return ok(await ctx.info.referralSummary({ user: resolveUser(ctx, user) }));
      } catch (e) {
        return toolError("get_referral_summary", e);
      }
    }
  );
}

// src/tools/account/getTradeHistory.ts
import { z as z11 } from "zod";
function registerGetTradeHistoryTool(server2, ctx) {
  server2.registerTool(
    "get_trade_history",
    {
      title: "Get trade history",
      description: "Fills / trade history. Omit user to use HOTSTUFF_ACCOUNT_ADDRESS.",
      inputSchema: z11.object({
        user: z11.string().optional(),
        instrumentId: z11.string().optional(),
        limit: z11.number().int().positive().optional()
      }),
      annotations: READ_ONLY_ANNOTATION
    },
    async ({ user, instrumentId, limit }) => {
      try {
        return ok(
          await ctx.info.tradeHistory({
            user: resolveUser(ctx, user),
            instrumentId,
            limit
          })
        );
      } catch (e) {
        return toolError("get_trade_history", e);
      }
    }
  );
}

// src/tools/account/getTransferHistory.ts
import { z as z12 } from "zod";
function registerGetTransferHistoryTool(server2, ctx) {
  server2.registerTool(
    "get_transfer_history",
    {
      title: "Get transfer history",
      description: "Deposits / withdrawals / transfers. Omit user to use HOTSTUFF_ACCOUNT_ADDRESS.",
      inputSchema: z12.object({
        user: z12.string().optional(),
        limit: z12.number().int().positive().optional()
      }),
      annotations: READ_ONLY_ANNOTATION
    },
    async ({ user, limit }) => {
      try {
        return ok(await ctx.info.transferHistory({ user: resolveUser(ctx, user), limit }));
      } catch (e) {
        return toolError("get_transfer_history", e);
      }
    }
  );
}

// src/tools/account/getUserFeeInfo.ts
import { z as z13 } from "zod";
function registerGetUserFeeInfoTool(server2, ctx) {
  server2.registerTool(
    "get_user_fee_info",
    {
      title: "Get user fee info",
      description: "Fee tier and volume. Omit user to use HOTSTUFF_ACCOUNT_ADDRESS.",
      inputSchema: z13.object({
        user: z13.string().optional()
      }),
      annotations: READ_ONLY_ANNOTATION
    },
    async ({ user }) => {
      try {
        return ok(await ctx.info.userFeeInfo({ user: resolveUser(ctx, user) }));
      } catch (e) {
        return toolError("get_user_fee_info", e);
      }
    }
  );
}

// src/tools/account/index.ts
function registerAccountTools(server2, ctx) {
  registerGetAccountSummaryTool(server2, ctx);
  registerGetAccountInfoTool(server2, ctx);
  registerGetOpenOrdersTool(server2, ctx);
  registerGetPositionsTool(server2, ctx);
  registerGetOrderHistoryTool(server2, ctx);
  registerGetTradeHistoryTool(server2, ctx);
  registerGetFundingHistoryTool(server2, ctx);
  registerGetTransferHistoryTool(server2, ctx);
  registerGetUserFeeInfoTool(server2, ctx);
  registerGetInstrumentLeverageTool(server2, ctx);
  registerGetReferralSummaryTool(server2, ctx);
  registerGetAgentsTool(server2, ctx);
  registerGetAccountHistoryTool(server2, ctx);
}

// src/tools/execute/addAgent.ts
import { privateKeyToAccount as privateKeyToAccount3 } from "viem/accounts";
import { z as z14 } from "zod";

// src/utils/requireExchange.ts
function requireExchange(ctx) {
  if (!ctx.exchange) {
    throw new Error(
      "HOTSTUFF_PRIVATE_KEY is required for this action. Set it in the MCP server environment."
    );
  }
  return ctx.exchange;
}

// src/tools/execute/addAgent.ts
function registerAddAgentTool(server2, ctx) {
  server2.registerTool(
    "add_agent",
    {
      title: "Add agent",
      description: "Register a trading agent. Requires HOTSTUFF_PRIVATE_KEY.",
      inputSchema: z14.object({
        agentName: z14.string().min(1),
        agentPrivateKey: z14.string().min(1),
        agent: z14.string().optional(),
        signer: z14.string().optional(),
        forAccount: z14.string().default(""),
        validUntil: z14.number().int().min(0).optional()
      }),
      annotations: DESTRUCTIVE_ANNOTATION
    },
    async ({ agentName, agentPrivateKey, agent, signer, forAccount, validUntil }) => {
      try {
        const ex = requireExchange(ctx);
        const derivedAgent = privateKeyToAccount3(agentPrivateKey).address;
        const resolvedSigner = signer ?? ctx.config.accountAddress;
        if (!resolvedSigner) {
          throw new Error(
            "add_agent requires signer: set HOTSTUFF_ACCOUNT_ADDRESS or pass signer."
          );
        }
        return ok(
          await ex.addAgent({
            agentName,
            agent: agent ?? derivedAgent,
            forAccount,
            agentPrivateKey,
            signer: resolvedSigner,
            validUntil: validUntil ?? expiresIn(30 * 24 * 3600)
          })
        );
      } catch (e) {
        return toolError("add_agent", e);
      }
    }
  );
}

// src/tools/execute/approveBrokerFee.ts
import { z as z15 } from "zod";
function registerApproveBrokerFeeTool(server2, ctx) {
  server2.registerTool(
    "approve_broker_fee",
    {
      title: "Approve broker fee",
      description: "Approve max broker fee rate. Requires HOTSTUFF_PRIVATE_KEY.",
      inputSchema: z15.object({
        broker: z15.string().min(1),
        maxFeeRate: z15.string().min(1)
      }),
      annotations: DESTRUCTIVE_ANNOTATION
    },
    async ({ broker, maxFeeRate }) => {
      try {
        const ex = requireExchange(ctx);
        return ok(await ex.approveBrokerFee({ broker, maxFeeRate }));
      } catch (e) {
        return toolError("approve_broker_fee", e);
      }
    }
  );
}

// src/tools/execute/cancelAllOrders.ts
import { z as z16 } from "zod";
function registerCancelAllOrdersTool(server2, ctx) {
  server2.registerTool(
    "cancel_all_orders",
    {
      title: "Cancel all orders",
      description: "Cancel all open orders for the signer. Requires HOTSTUFF_PRIVATE_KEY.",
      inputSchema: z16.object({
        expiresAfter: z16.number().int().min(0).optional()
      }),
      annotations: DESTRUCTIVE_ANNOTATION
    },
    async ({ expiresAfter }) => {
      try {
        const ex = requireExchange(ctx);
        return ok(await ex.cancelAll({ expiresAfter: expiresAfter ?? expiresIn(3600) }));
      } catch (e) {
        return toolError("cancel_all_orders", e);
      }
    }
  );
}

// src/tools/execute/cancelOrdersByClientId.ts
import { z as z17 } from "zod";
function registerCancelOrdersByClientIdTool(server2, ctx) {
  server2.registerTool(
    "cancel_orders_by_client_id",
    {
      title: "Cancel orders by client id",
      description: "Cancel by client order id. Requires HOTSTUFF_PRIVATE_KEY.",
      inputSchema: z17.object({
        cancels: z17.array(
          z17.object({
            cloid: z17.string().min(1),
            instrumentId: z17.number().int().positive()
          })
        ).min(1),
        expiresAfter: z17.number().int().min(0).optional()
      }),
      annotations: DESTRUCTIVE_ANNOTATION
    },
    async ({ cancels, expiresAfter }) => {
      try {
        const ex = requireExchange(ctx);
        return ok(
          await ex.cancelByCloid({
            cancels,
            expiresAfter: expiresAfter ?? expiresIn(3600)
          })
        );
      } catch (e) {
        return toolError("cancel_orders_by_client_id", e);
      }
    }
  );
}

// src/tools/execute/cancelOrdersById.ts
import { z as z18 } from "zod";
function registerCancelOrdersByIdTool(server2, ctx) {
  server2.registerTool(
    "cancel_orders_by_id",
    {
      title: "Cancel orders by id",
      description: "Cancel by Hotstuff order id. Requires HOTSTUFF_PRIVATE_KEY.",
      inputSchema: z18.object({
        cancels: z18.array(
          z18.object({
            oid: z18.number().int().min(0),
            instrumentId: z18.number().int().positive()
          })
        ).min(1),
        expiresAfter: z18.number().int().min(0).optional()
      }),
      annotations: DESTRUCTIVE_ANNOTATION
    },
    async ({ cancels, expiresAfter }) => {
      try {
        const ex = requireExchange(ctx);
        return ok(
          await ex.cancelByOid({
            cancels,
            expiresAfter: expiresAfter ?? expiresIn(3600)
          })
        );
      } catch (e) {
        return toolError("cancel_orders_by_id", e);
      }
    }
  );
}

// src/tools/execute/claimReferralRewards.ts
import { z as z19 } from "zod";
function registerClaimReferralRewardsTool(server2, ctx) {
  server2.registerTool(
    "claim_referral_rewards",
    {
      title: "Claim referral rewards",
      description: "Claim referral rewards. Requires HOTSTUFF_PRIVATE_KEY.",
      inputSchema: z19.object({
        collateralId: z19.number().int().positive(),
        spot: z19.boolean()
      }),
      annotations: DESTRUCTIVE_ANNOTATION
    },
    async ({ collateralId, spot }) => {
      try {
        const ex = requireExchange(ctx);
        return ok(await ex.claimReferralRewards({ collateralId, spot }));
      } catch (e) {
        return toolError("claim_referral_rewards", e);
      }
    }
  );
}

// src/tools/execute/createReferralCode.ts
import { z as z20 } from "zod";
function registerCreateReferralCodeTool(server2, ctx) {
  server2.registerTool(
    "create_referral_code",
    {
      title: "Create referral code",
      description: "Create a referral code. Requires HOTSTUFF_PRIVATE_KEY.",
      inputSchema: z20.object({
        code: z20.string().min(1)
      }),
      annotations: DESTRUCTIVE_ANNOTATION
    },
    async ({ code }) => {
      try {
        const ex = requireExchange(ctx);
        return ok(await ex.createReferralCode({ code }));
      } catch (e) {
        return toolError("create_referral_code", e);
      }
    }
  );
}

// src/tools/execute/depositToVault.ts
import { z as z21 } from "zod";
function registerDepositToVaultTool(server2, ctx) {
  server2.registerTool(
    "deposit_to_vault",
    {
      title: "Deposit to vault",
      description: "Deposit into a vault. Requires HOTSTUFF_PRIVATE_KEY.",
      inputSchema: z21.object({
        vaultAddress: z21.string().min(1),
        amount: z21.string().min(1)
      }),
      annotations: DESTRUCTIVE_ANNOTATION
    },
    async ({ vaultAddress, amount }) => {
      try {
        const ex = requireExchange(ctx);
        return ok(await ex.depositToVault({ vaultAddress, amount }));
      } catch (e) {
        return toolError("deposit_to_vault", e);
      }
    }
  );
}

// src/tools/execute/placeOrder.ts
import { z as z23 } from "zod";

// src/tools/execute/schemas.ts
import { z as z22 } from "zod";
var orderUnitSchema = z22.object({
  instrumentId: z22.number().int().positive(),
  side: z22.enum(["b", "s"]),
  positionSide: z22.enum(["LONG", "SHORT", "BOTH"]),
  price: z22.string().min(1),
  size: z22.string().min(1),
  tif: z22.enum(["GTC", "IOC", "FOK"]),
  ro: z22.boolean(),
  po: z22.boolean(),
  cloid: z22.string().optional(),
  triggerPx: z22.string().optional(),
  isMarket: z22.boolean().optional(),
  tpsl: z22.enum(["", "tp", "sl"]).optional(),
  grouping: z22.enum(["", "position", "normal"]).optional()
});

// src/tools/execute/placeOrder.ts
function registerPlaceOrderTool(server2, ctx) {
  server2.registerTool(
    "place_order",
    {
      title: "Place order",
      description: "Place signed orders. Requires HOTSTUFF_PRIVATE_KEY. Use get_instruments for instrumentId.",
      inputSchema: z23.object({
        orders: z23.array(orderUnitSchema).min(1),
        brokerConfig: z23.object({
          broker: z23.string(),
          fee: z23.string()
        }).optional(),
        expiresAfter: z23.number().int().min(0).optional()
      }),
      annotations: DESTRUCTIVE_ANNOTATION
    },
    async ({ orders, brokerConfig, expiresAfter }) => {
      try {
        const ex = requireExchange(ctx);
        const resolvedOrders = orders.map((o, i) => ({
          ...o,
          cloid: o.cloid ?? `mcp-${Date.now()}-${i}`
        }));
        return ok(
          await ex.placeOrder({
            orders: resolvedOrders,
            brokerConfig: brokerConfig ? { broker: brokerConfig.broker, fee: brokerConfig.fee } : void 0,
            expiresAfter: expiresAfter ?? expiresIn(3600)
          })
        );
      } catch (e) {
        return toolError("place_order", e);
      }
    }
  );
}

// src/tools/execute/redeemFromVault.ts
import { z as z24 } from "zod";
function registerRedeemFromVaultTool(server2, ctx) {
  server2.registerTool(
    "redeem_from_vault",
    {
      title: "Redeem from vault",
      description: "Redeem vault shares. Requires HOTSTUFF_PRIVATE_KEY.",
      inputSchema: z24.object({
        vaultAddress: z24.string().min(1),
        shares: z24.string().min(1)
      }),
      annotations: DESTRUCTIVE_ANNOTATION
    },
    async ({ vaultAddress, shares }) => {
      try {
        const ex = requireExchange(ctx);
        return ok(
          await ex.redeemFromVault({ vaultAddress, shares })
        );
      } catch (e) {
        return toolError("redeem_from_vault", e);
      }
    }
  );
}

// src/tools/execute/requestDerivativeWithdraw.ts
import { z as z25 } from "zod";
function registerRequestDerivativeWithdrawTool(server2, ctx) {
  server2.registerTool(
    "request_derivative_withdraw",
    {
      title: "Request derivative withdraw",
      description: "Derivatives collateral withdrawal. Requires HOTSTUFF_PRIVATE_KEY.",
      inputSchema: z25.object({
        collateralId: z25.number().int().positive(),
        amount: z25.string().min(1),
        chainId: z25.number().int().positive()
      }),
      annotations: DESTRUCTIVE_ANNOTATION
    },
    async ({ collateralId, amount, chainId }) => {
      try {
        const ex = requireExchange(ctx);
        return ok(await ex.accountDerivativeWithdrawRequest({ collateralId, amount, chainId }));
      } catch (e) {
        return toolError("request_derivative_withdraw", e);
      }
    }
  );
}

// src/tools/execute/requestSpotWithdraw.ts
import { z as z26 } from "zod";
function registerRequestSpotWithdrawTool(server2, ctx) {
  server2.registerTool(
    "request_spot_withdraw",
    {
      title: "Request spot withdraw",
      description: "Spot collateral withdrawal to external chain. Requires HOTSTUFF_PRIVATE_KEY.",
      inputSchema: z26.object({
        collateralId: z26.number().int().positive(),
        amount: z26.string().min(1),
        chainId: z26.number().int().positive()
      }),
      annotations: DESTRUCTIVE_ANNOTATION
    },
    async ({ collateralId, amount, chainId }) => {
      try {
        const ex = requireExchange(ctx);
        return ok(await ex.accountSpotWithdrawRequest({ collateralId, amount, chainId }));
      } catch (e) {
        return toolError("request_spot_withdraw", e);
      }
    }
  );
}

// src/tools/execute/revokeAgent.ts
import { z as z27 } from "zod";
function registerRevokeAgentTool(server2, ctx) {
  server2.registerTool(
    "revoke_agent",
    {
      title: "Revoke agent",
      description: "Revoke a trading agent. Requires HOTSTUFF_PRIVATE_KEY.",
      inputSchema: z27.object({
        agent: z27.string().min(1),
        forAccount: z27.string().optional()
      }),
      annotations: DESTRUCTIVE_ANNOTATION
    },
    async ({ agent, forAccount }) => {
      try {
        const ex = requireExchange(ctx);
        return ok(await ex.revokeAgent({ agent, forAccount }));
      } catch (e) {
        return toolError("revoke_agent", e);
      }
    }
  );
}

// src/tools/execute/setReferrer.ts
import { z as z28 } from "zod";
function registerSetReferrerTool(server2, ctx) {
  server2.registerTool(
    "set_referrer",
    {
      title: "Set referrer",
      description: "Bind referrer via code. Requires HOTSTUFF_PRIVATE_KEY.",
      inputSchema: z28.object({
        code: z28.string().min(1)
      }),
      annotations: DESTRUCTIVE_ANNOTATION
    },
    async ({ code }) => {
      try {
        const ex = requireExchange(ctx);
        return ok(await ex.setReferrer({ code }));
      } catch (e) {
        return toolError("set_referrer", e);
      }
    }
  );
}

// src/tools/execute/transferDerivativeBalance.ts
import { z as z29 } from "zod";
function registerTransferDerivativeBalanceTool(server2, ctx) {
  server2.registerTool(
    "transfer_derivative_balance",
    {
      title: "Transfer derivative balance",
      description: "Transfer derivatives balance to another address. Requires HOTSTUFF_PRIVATE_KEY.",
      inputSchema: z29.object({
        collateralId: z29.number().int().positive(),
        amount: z29.string().min(1),
        destination: z29.string().min(1)
      }),
      annotations: DESTRUCTIVE_ANNOTATION
    },
    async ({ collateralId, amount, destination }) => {
      try {
        const ex = requireExchange(ctx);
        return ok(
          await ex.accountDerivativeBalanceTransferRequest({
            collateralId,
            amount,
            destination
          })
        );
      } catch (e) {
        return toolError("transfer_derivative_balance", e);
      }
    }
  );
}

// src/tools/execute/transferInternalBalance.ts
import { z as z30 } from "zod";
function registerTransferInternalBalanceTool(server2, ctx) {
  server2.registerTool(
    "transfer_internal_balance",
    {
      title: "Transfer internal balance",
      description: "Move balance between spot and derivatives for the signer. Requires HOTSTUFF_PRIVATE_KEY.",
      inputSchema: z30.object({
        collateralId: z30.number().int().positive(),
        amount: z30.string().min(1),
        toDerivativesAccount: z30.boolean()
      }),
      annotations: DESTRUCTIVE_ANNOTATION
    },
    async ({ collateralId, amount, toDerivativesAccount }) => {
      try {
        const ex = requireExchange(ctx);
        return ok(
          await ex.accountInternalBalanceTransferRequest({
            collateralId,
            amount,
            toDerivativesAccount
          })
        );
      } catch (e) {
        return toolError("transfer_internal_balance", e);
      }
    }
  );
}

// src/tools/execute/transferSpotBalance.ts
import { z as z31 } from "zod";
function registerTransferSpotBalanceTool(server2, ctx) {
  server2.registerTool(
    "transfer_spot_balance",
    {
      title: "Transfer spot balance",
      description: "Transfer spot balance to another address. Requires HOTSTUFF_PRIVATE_KEY.",
      inputSchema: z31.object({
        collateralId: z31.number().int().positive(),
        amount: z31.string().min(1),
        destination: z31.string().min(1)
      }),
      annotations: DESTRUCTIVE_ANNOTATION
    },
    async ({ collateralId, amount, destination }) => {
      try {
        const ex = requireExchange(ctx);
        return ok(
          await ex.accountSpotBalanceTransferRequest({
            collateralId,
            amount,
            destination
          })
        );
      } catch (e) {
        return toolError("transfer_spot_balance", e);
      }
    }
  );
}

// src/tools/execute/updatePerpLeverage.ts
import { z as z32 } from "zod";
function registerUpdatePerpLeverageTool(server2, ctx) {
  server2.registerTool(
    "update_perp_leverage",
    {
      title: "Update perp leverage",
      description: "Update leverage for a perpetual. Requires HOTSTUFF_PRIVATE_KEY.",
      inputSchema: z32.object({
        instrumentId: z32.number().int().positive(),
        leverage: z32.number().int().positive()
      }),
      annotations: DESTRUCTIVE_ANNOTATION
    },
    async ({ instrumentId, leverage }) => {
      try {
        const ex = requireExchange(ctx);
        return ok(await ex.updatePerpInstrumentLeverage({ instrumentId, leverage }));
      } catch (e) {
        return toolError("update_perp_leverage", e);
      }
    }
  );
}

// src/tools/execute/index.ts
function registerExecuteTools(server2, ctx) {
  registerPlaceOrderTool(server2, ctx);
  registerCancelOrdersByIdTool(server2, ctx);
  registerCancelOrdersByClientIdTool(server2, ctx);
  registerCancelAllOrdersTool(server2, ctx);
  registerUpdatePerpLeverageTool(server2, ctx);
  registerAddAgentTool(server2, ctx);
  registerRevokeAgentTool(server2, ctx);
  registerApproveBrokerFeeTool(server2, ctx);
  registerCreateReferralCodeTool(server2, ctx);
  registerSetReferrerTool(server2, ctx);
  registerClaimReferralRewardsTool(server2, ctx);
  registerRequestSpotWithdrawTool(server2, ctx);
  registerRequestDerivativeWithdrawTool(server2, ctx);
  registerTransferSpotBalanceTool(server2, ctx);
  registerTransferDerivativeBalanceTool(server2, ctx);
  registerTransferInternalBalanceTool(server2, ctx);
  registerDepositToVaultTool(server2, ctx);
  registerRedeemFromVaultTool(server2, ctx);
}

// src/tools/explorer/getBlockDetails.ts
import { z as z33 } from "zod";
function registerGetBlockDetailsTool(server2, ctx) {
  server2.registerTool(
    "get_block_details",
    {
      title: "Get block details",
      description: "Block by hash or height.",
      inputSchema: z33.object({
        block_hash: z33.string().optional(),
        block_height: z33.number().int().min(0).optional()
      }),
      annotations: READ_ONLY_ANNOTATION
    },
    async ({ block_hash, block_height }) => {
      try {
        return ok(await ctx.info.blockDetails({ block_hash, block_height }));
      } catch (e) {
        return toolError("get_block_details", e);
      }
    }
  );
}

// src/tools/explorer/getBlocks.ts
import { z as z34 } from "zod";
function registerGetBlocksTool(server2, ctx) {
  server2.registerTool(
    "get_blocks",
    {
      title: "Get blocks",
      description: "Recent chain blocks.",
      inputSchema: z34.object({
        offset: z34.number().int().min(0).default(0),
        limit: z34.number().int().positive().default(10)
      }),
      annotations: READ_ONLY_ANNOTATION
    },
    async ({ offset, limit }) => {
      try {
        return ok(await ctx.info.blocks({ offset, limit }));
      } catch (e) {
        return toolError("get_blocks", e);
      }
    }
  );
}

// src/tools/explorer/getTransactionDetails.ts
import { z as z35 } from "zod";
function registerGetTransactionDetailsTool(server2, ctx) {
  server2.registerTool(
    "get_transaction_details",
    {
      title: "Get transaction details",
      description: "Transaction by hash.",
      inputSchema: z35.object({
        tx_hash: z35.string().min(1)
      }),
      annotations: READ_ONLY_ANNOTATION
    },
    async ({ tx_hash }) => {
      try {
        return ok(await ctx.info.transactionDetails({ tx_hash }));
      } catch (e) {
        return toolError("get_transaction_details", e);
      }
    }
  );
}

// src/tools/explorer/getTransactions.ts
import { z as z36 } from "zod";
function registerGetTransactionsTool(server2, ctx) {
  server2.registerTool(
    "get_transactions",
    {
      title: "Get transactions",
      description: "Chain transactions with optional filters.",
      inputSchema: z36.object({
        offset: z36.number().int().min(0).optional(),
        limit: z36.number().int().positive().optional(),
        account: z36.string().optional(),
        txType: z36.number().int().min(0).optional()
      }),
      annotations: READ_ONLY_ANNOTATION
    },
    async ({ offset, limit, account, txType }) => {
      try {
        return ok(
          await ctx.info.transactions({
            offset,
            limit,
            filter: account || txType !== void 0 ? { account, tx_type: txType } : void 0
          })
        );
      } catch (e) {
        return toolError("get_transactions", e);
      }
    }
  );
}

// src/tools/explorer/index.ts
function registerExplorerTools(server2, ctx) {
  registerGetBlocksTool(server2, ctx);
  registerGetBlockDetailsTool(server2, ctx);
  registerGetTransactionsTool(server2, ctx);
  registerGetTransactionDetailsTool(server2, ctx);
}

// src/tools/market/getBestBidOffer.ts
import { z as z37 } from "zod";
function registerGetBestBidOfferTool(server2, ctx) {
  server2.registerTool(
    "get_best_bid_offer",
    {
      title: "Get best bid and offer",
      description: 'Best bid/ask for a market. Example: { "symbol": "BTC-PERP" }',
      inputSchema: z37.object({
        symbol: z37.string().min(1).describe("Market symbol.")
      }),
      annotations: READ_ONLY_ANNOTATION
    },
    async ({ symbol }) => {
      try {
        return ok(await ctx.info.bbo({ symbol }));
      } catch (e) {
        return toolError("get_best_bid_offer", e);
      }
    }
  );
}

// src/tools/market/getChart.ts
import { z as z38 } from "zod";
function registerGetChartTool(server2, ctx) {
  server2.registerTool(
    "get_chart",
    {
      title: "Get chart",
      description: 'OHLCV candles. Example: { "symbol": "BTC-PERP", "resolution": "60", "chart_type": "mark", "from": 1700000000, "to": 1700086400 }',
      inputSchema: z38.object({
        symbol: z38.string().min(1).describe("Market symbol."),
        resolution: z38.enum(["1", "5", "15", "60", "240", "1D", "1W"]).describe("Candle resolution."),
        chart_type: z38.enum(["mark", "ltp", "index"]).describe("Price series."),
        from: z38.number().int().min(0).describe("Start unix seconds."),
        to: z38.number().int().min(0).describe("End unix seconds.")
      }),
      annotations: READ_ONLY_ANNOTATION
    },
    async ({ symbol, resolution, chart_type, from, to }) => {
      try {
        return ok(await ctx.info.chart({ symbol, resolution, chart_type, from, to }));
      } catch (e) {
        return toolError("get_chart", e);
      }
    }
  );
}

// src/tools/market/getInstruments.ts
import { z as z39 } from "zod";
function registerGetInstrumentsTool(server2, ctx) {
  server2.registerTool(
    "get_instruments",
    {
      title: "Get instruments",
      description: 'List all Hotstuff trading instruments (perps, spot, or all). Use before placing orders. Example: { "type": "perps" }',
      inputSchema: z39.object({
        type: z39.enum(["perps", "spot", "all"]).default("all").describe("Filter: perps, spot, or all.")
      }),
      annotations: READ_ONLY_ANNOTATION
    },
    async ({ type }) => {
      try {
        return ok(await ctx.info.instruments({ type }));
      } catch (e) {
        return toolError("get_instruments", e);
      }
    }
  );
}

// src/tools/market/getMidPrices.ts
import { z as z40 } from "zod";
function normalizeSymbol(symbol) {
  if (!symbol) return void 0;
  const normalized = symbol.trim().toUpperCase();
  return normalized.length > 0 ? normalized : void 0;
}
function isRecord(value) {
  return typeof value === "object" && value !== null;
}
function parseMidPrice(value) {
  if (!isRecord(value)) return null;
  if (typeof value.symbol !== "string") return null;
  if (typeof value.mid_price !== "string") return null;
  return {
    symbol: value.symbol,
    mid_price: value.mid_price
  };
}
function pickMidFromResponse(response, requestedSymbol) {
  if (Array.isArray(response)) {
    const mids = response.map(parseMidPrice).filter((item) => item !== null);
    if (mids.length === 0) return null;
    return mids.find((item) => item.symbol.toUpperCase() === requestedSymbol) ?? mids[0];
  }
  return parseMidPrice(response);
}
function hasSymbolRequiredError(error) {
  const message = error instanceof Error ? error.message : String(error);
  return message.toLowerCase().includes("symbol is required");
}
async function getSingleMidPrice(ctx, symbol) {
  try {
    const midsResponse = await ctx.info.mids({ symbol });
    const midFromMids = pickMidFromResponse(midsResponse, symbol);
    if (midFromMids) return midFromMids;
  } catch {
  }
  try {
    const ticker = await ctx.info.ticker({ symbol });
    const first = ticker[0];
    if (first && typeof first.symbol === "string" && typeof first.mid_price === "string") {
      return { symbol: first.symbol, mid_price: first.mid_price };
    }
  } catch {
  }
  return null;
}
async function getAllMidPricesWithFallback(ctx, originalError) {
  const instruments = await ctx.info.instruments({ type: "all" });
  const symbols = Array.from(
    new Set(
      [
        ...instruments.perps.map((instrument) => instrument.name),
        ...instruments.spot.map((instrument) => instrument.name)
      ].map((symbol) => normalizeSymbol(symbol)).filter((symbol) => Boolean(symbol))
    )
  );
  const mids = [];
  for (const symbol of symbols) {
    const mid = await getSingleMidPrice(ctx, symbol);
    if (mid) mids.push(mid);
  }
  if (mids.length === 0) {
    const reason = originalError instanceof Error ? originalError.message : String(originalError);
    throw new Error(`Could not aggregate mid prices after mids() rejected empty input: ${reason}`);
  }
  return mids;
}
function registerGetMidPricesTool(server2, ctx) {
  server2.registerTool(
    "get_mid_prices",
    {
      title: "Get mid prices",
      description: 'Mid prices. Use without params for all symbols; pass `symbol` to fetch one market (e.g. { "symbol": "BTC-PERP" }).',
      inputSchema: z40.object({
        symbol: z40.string().trim().min(1).optional().describe("Optional market symbol, e.g. BTC-PERP.")
      }),
      annotations: READ_ONLY_ANNOTATION
    },
    async ({ symbol }) => {
      try {
        const normalizedSymbol = normalizeSymbol(symbol);
        if (normalizedSymbol) {
          const mid = await getSingleMidPrice(ctx, normalizedSymbol);
          if (!mid) {
            throw new Error(`No mid price found for symbol: ${normalizedSymbol}`);
          }
          return ok([mid]);
        }
        try {
          return ok(await ctx.info.mids({}));
        } catch (e) {
          if (!hasSymbolRequiredError(e)) {
            throw e;
          }
          return ok(await getAllMidPricesWithFallback(ctx, e));
        }
      } catch (e) {
        return toolError("get_mid_prices", e);
      }
    }
  );
}

// src/tools/market/getOraclePrice.ts
import { z as z41 } from "zod";
function buildOracleSymbolCandidates(rawSymbol) {
  const normalized = rawSymbol.trim().toUpperCase();
  if (!normalized.includes("/")) {
    return [normalized, `${normalized}/USDC`];
  }
  return [normalized];
}
function registerGetOraclePriceTool(server2, ctx) {
  server2.registerTool(
    "get_oracle_price",
    {
      title: "Get oracle price",
      description: "Oracle / index price for a symbol. Accepts asset symbols (`BTC`) and pair symbols (`BTC/USDC`).",
      inputSchema: z41.object({
        symbol: z41.string().trim().min(1).describe("Asset or pair symbol, e.g. BTC or BTC/USDC.")
      }),
      annotations: READ_ONLY_ANNOTATION
    },
    async ({ symbol }) => {
      try {
        const candidates = buildOracleSymbolCandidates(symbol);
        let lastError;
        for (const candidate of candidates) {
          try {
            return ok(await ctx.info.oracle({ symbol: candidate }));
          } catch (e) {
            lastError = e;
          }
        }
        const reason = lastError instanceof Error ? lastError.message : String(lastError);
        throw new Error(`${reason}. Tried symbol candidates: ${candidates.join(", ")}`);
      } catch (e) {
        return toolError("get_oracle_price", e);
      }
    }
  );
}

// src/tools/market/getOrderbook.ts
import { z as z42 } from "zod";
function registerGetOrderbookTool(server2, ctx) {
  server2.registerTool(
    "get_orderbook",
    {
      title: "Get orderbook",
      description: 'Orderbook snapshot. Example: { "symbol": "BTC-PERP", "depth": 20 }',
      inputSchema: z42.object({
        symbol: z42.string().min(1).describe("Market symbol."),
        depth: z42.number().int().positive().optional().describe("Levels per side.")
      }),
      annotations: READ_ONLY_ANNOTATION
    },
    async ({ symbol, depth }) => {
      try {
        return ok(await ctx.info.orderbook({ symbol, depth }));
      } catch (e) {
        return toolError("get_orderbook", e);
      }
    }
  );
}

// src/tools/market/getRecentTrades.ts
import { z as z43 } from "zod";
function registerGetRecentTradesTool(server2, ctx) {
  server2.registerTool(
    "get_recent_trades",
    {
      title: "Get recent trades",
      description: 'Recent public trades. Example: { "symbol": "BTC-PERP", "limit": 50 }',
      inputSchema: z43.object({
        symbol: z43.string().min(1).describe("Market symbol."),
        limit: z43.number().int().positive().optional().describe("Max trades.")
      }),
      annotations: READ_ONLY_ANNOTATION
    },
    async ({ symbol, limit }) => {
      try {
        return ok(await ctx.info.trades({ symbol, limit }));
      } catch (e) {
        return toolError("get_recent_trades", e);
      }
    }
  );
}

// src/tools/market/getSupportedCollateral.ts
import { z as z44 } from "zod";
function registerGetSupportedCollateralTool(server2, ctx) {
  server2.registerTool(
    "get_supported_collateral",
    {
      title: "Get supported collateral",
      description: "List collateral assets with decimals and bridge configuration.",
      inputSchema: z44.object({}),
      annotations: READ_ONLY_ANNOTATION
    },
    async () => {
      try {
        return ok(await ctx.info.supportedCollateral({}));
      } catch (e) {
        return toolError("get_supported_collateral", e);
      }
    }
  );
}

// src/tools/market/getTicker.ts
import { z as z45 } from "zod";
function registerGetTickerTool(server2, ctx) {
  server2.registerTool(
    "get_ticker",
    {
      title: "Get ticker",
      description: 'Ticker for a market (mark, funding, 24h volume). Example: { "symbol": "BTC-PERP" }',
      inputSchema: z45.object({
        symbol: z45.string().min(1).describe("Market symbol, e.g. BTC-PERP.")
      }),
      annotations: READ_ONLY_ANNOTATION
    },
    async ({ symbol }) => {
      try {
        return ok(await ctx.info.ticker({ symbol }));
      } catch (e) {
        return toolError("get_ticker", e);
      }
    }
  );
}

// src/tools/market/index.ts
function registerMarketTools(server2, ctx) {
  registerGetInstrumentsTool(server2, ctx);
  registerGetSupportedCollateralTool(server2, ctx);
  registerGetOraclePriceTool(server2, ctx);
  registerGetTickerTool(server2, ctx);
  registerGetMidPricesTool(server2, ctx);
  registerGetBestBidOfferTool(server2, ctx);
  registerGetOrderbookTool(server2, ctx);
  registerGetRecentTradesTool(server2, ctx);
  registerGetChartTool(server2, ctx);
}

// src/tools/stream/collectTradesStream.ts
import { z as z46 } from "zod";
var MAX_SECONDS = 120;
var MAX_EVENTS = 2e3;
function registerCollectTradesStreamTool(server2, ctx) {
  server2.registerTool(
    "collect_trades_stream",
    {
      title: "Collect trades stream",
      description: `Subscribe to public trades for an instrument for a bounded time (max ${MAX_SECONDS}s). Pass instrument_id from get_instruments (string).`,
      inputSchema: z46.object({
        instrument_id: z46.string().min(1).describe("Instrument id string from get_instruments."),
        seconds: z46.number().int().min(1).max(MAX_SECONDS).default(10),
        max_events: z46.number().int().min(1).max(MAX_EVENTS).default(300)
      }),
      annotations: READ_ONLY_ANNOTATION
    },
    async ({ instrument_id, seconds, max_events }) => {
      try {
        const events = [];
        const sub = await ctx.subscription.trade(
          { instrumentId: instrument_id },
          (ev) => {
            if (events.length < max_events) {
              events.push(ev.detail);
            }
          }
        );
        await new Promise((r) => setTimeout(r, seconds * 1e3));
        await sub.unsubscribe();
        return ok({
          instrument_id,
          seconds,
          count: events.length,
          events
        });
      } catch (e) {
        return toolError("collect_trades_stream", e);
      }
    }
  );
}

// src/tools/stream/index.ts
function registerStreamTools(server2, ctx) {
  registerCollectTradesStreamTool(server2, ctx);
}

// src/tools/vault/getSubVaults.ts
import { z as z47 } from "zod";
function registerGetSubVaultsTool(server2, ctx) {
  server2.registerTool(
    "get_sub_vaults",
    {
      title: "Get sub-vaults",
      description: "Sub-vaults for a vault address.",
      inputSchema: z47.object({
        vaultAddress: z47.string().min(1).describe("Vault address.")
      }),
      annotations: READ_ONLY_ANNOTATION
    },
    async ({ vaultAddress }) => {
      try {
        return ok(await ctx.info.subVaults({ vaultAddress }));
      } catch (e) {
        return toolError("get_sub_vaults", e);
      }
    }
  );
}

// src/tools/vault/getVaultBalances.ts
import { z as z48 } from "zod";
function registerGetVaultBalancesTool(server2, ctx) {
  server2.registerTool(
    "get_vault_balances",
    {
      title: "Get vault balances",
      description: "Balances for a vault; optional user filter.",
      inputSchema: z48.object({
        vaultAddress: z48.string().min(1),
        user: z48.string().optional()
      }),
      annotations: READ_ONLY_ANNOTATION
    },
    async ({ vaultAddress, user }) => {
      try {
        return ok(
          await ctx.info.vaultBalances({
            vaultAddress,
            user
          })
        );
      } catch (e) {
        return toolError("get_vault_balances", e);
      }
    }
  );
}

// src/tools/vault/getVaults.ts
import { z as z49 } from "zod";
function registerGetVaultsTool(server2, ctx) {
  server2.registerTool(
    "get_vaults",
    {
      title: "Get vaults",
      description: "List all Hotstuff vaults.",
      inputSchema: z49.object({}),
      annotations: READ_ONLY_ANNOTATION
    },
    async () => {
      try {
        return ok(await ctx.info.vaults({}));
      } catch (e) {
        return toolError("get_vaults", e);
      }
    }
  );
}

// src/tools/vault/index.ts
function registerVaultTools(server2, ctx) {
  registerGetVaultsTool(server2, ctx);
  registerGetSubVaultsTool(server2, ctx);
  registerGetVaultBalancesTool(server2, ctx);
}

// src/tools/index.ts
function registerAllTools(server2, ctx) {
  registerMarketTools(server2, ctx);
  registerAccountTools(server2, ctx);
  registerVaultTools(server2, ctx);
  registerExplorerTools(server2, ctx);
  registerExecuteTools(server2, ctx);
  registerStreamTools(server2, ctx);
}

// src/server.ts
async function createServer() {
  const config = loadConfig();
  const ctx = createHotstuffContext(config);
  const server2 = new McpServer(
    { name: "Hotstuff MCP", version: "0.1.0" },
    { instructions: SERVER_INSTRUCTIONS }
  );
  registerAllTools(server2, ctx);
  registerAllResources(server2, ctx);
  return server2;
}

// src/index.ts
var server = await createServer();
var transport = new StdioServerTransport();
await server.connect(transport);
