// @hotstuff-labs/ts-sdk/dist/utils/transports.util.js
export const ENDPOINTS_URLS = {
    mainnet: {
        api: "https://testnet-api.hotstuff.trade/",
        rpc: "https://testnet-api.hotstuff.trade/",
        ws: "wss://testnet-api.hotstuff.trade/ws/",
    },
    testnet: {
        api: "https://testnet-api.hotstuff.trade/",
        rpc: "https://testnet-api.hotstuff.trade/",
        ws: "wss://testnet-api.hotstuff.trade/ws/",
    },
};
export function mergeRequestInit(...inits) {
    const merged = inits.reduce((acc, init) => ({ ...acc, ...init }), {});
    const headersList = inits
        .map((init) => init.headers)
        .filter((headers) => typeof headers === "object");
    if (headersList.length > 0) {
        merged.headers = mergeHeadersInit(...headersList);
    }
    const signals = inits
        .map((init) => init.signal)
        .filter((signal) => signal instanceof AbortSignal);
    if (signals.length > 0) {
        merged.signal = signals.length > 1 ? AbortSignal.any(signals) : signals[0];
    }
    return merged;
}
function mergeHeadersInit(...inits) {
    if (inits.length === 0 || inits.length === 1) {
        return new Headers(inits[0]);
    }
    const merged = new Headers();
    for (const headers of inits) {
        if (Symbol.iterator in headers) {
            for (const [key, value] of headers) {
                merged.set(key, value);
            }
        }
        else {
            for (const key in headers) {
                if (headers.hasOwnProperty(key)) {
                    merged.set(key, headers[key]);
                }
            }
        }
    }
    return merged;
}
