// @hotstuff-labs/ts-sdk/dist/types/transports/websocket.type.js
export var WSMethod;
(function (WSMethod) {
    WSMethod["SUBSCRIBE"] = "subscribe";
    WSMethod["UNSUBSCRIBE"] = "unsubscribe";
    WSMethod["PING"] = "ping";
    WSMethod["POST"] = "post";
})(WSMethod || (WSMethod = {}));
export var WSEventType;
(function (WSEventType) {
    WSEventType["DEPOSIT"] = "deposit";
    WSEventType["SPOT_WITHDRAW"] = "spot_withdraw";
    WSEventType["DERIVATIVE_WITHDRAW"] = "derivative_withdraw";
    WSEventType["ORDER_UPDATE"] = "order_update";
    WSEventType["TRADE"] = "trade";
    WSEventType["SPOT_TRADE"] = "spot_trade";
    WSEventType["ACCOUNT_BALANCE_UPDATED"] = "account_balance_updated";
    WSEventType["ACCOUNT_SPOT_BALANCE_UPDATED"] = "account_spot_balance_updated";
    WSEventType["AGENT_REGISTERED"] = "agent_registered";
    WSEventType["ACCOUNT_UPDATED"] = "account_updated";
    WSEventType["PRICE_INDEX_ADDED"] = "price_index_added";
    WSEventType["ORACLE_FEED_UPDATED"] = "oracle_feed_updated";
    WSEventType["PERP_LISTED"] = "perp_listed";
    WSEventType["SPOT_LISTED"] = "spot_listed";
    WSEventType["OPTION_LISTED"] = "option_listed";
    WSEventType["SPOT_BALANCE_TRANSFER"] = "spot_balance_transfer";
    WSEventType["DERIVATIVE_BALANCE_TRANSFER"] = "derivative_balance_transfer";
    WSEventType["INTERNAL_BALANCE_TRANSFER"] = "internal_balance_transfer";
    WSEventType["SPOT_SWAP_TRIGGERED"] = "spot_swap_triggered";
    WSEventType["EVM_COLLATERAL_ADDED"] = "evm_collateral_added";
    WSEventType["INDEXER_BOOK_UPDATE"] = "indexer_book_update";
    WSEventType["TICKER"] = "ticker";
})(WSEventType || (WSEventType = {}));
