// @hotstuff-labs/ts-sdk/dist/transports/http.transport.js
import { ENDPOINTS_URLS, mergeRequestInit } from '../utils';
export class HttpTransport {
    constructor(options) {
        this.isTestnet = options?.isTestnet ?? false;
        this.timeout = options?.timeout ?? 3000;
        this.server = {
            mainnet: {
                api: options?.server?.mainnet?.api ?? ENDPOINTS_URLS.mainnet.api,
                rpc: options?.server?.mainnet?.rpc ?? ENDPOINTS_URLS.mainnet.rpc,
            },
            testnet: {
                api: options?.server?.testnet?.api ?? ENDPOINTS_URLS.testnet.api,
                rpc: options?.server?.testnet?.rpc ?? ENDPOINTS_URLS.testnet.rpc,
            },
        };
        this.fetchOptions = options?.fetchOptions ?? {};
        this.onRequest = options?.onRequest;
        this.onResponse = options?.onResponse;
    }
    async request(endpoint, payload, signal, method = 'POST') {
        try {
            const url = new URL(endpoint, this.server[this.isTestnet ? 'testnet' : 'mainnet'][endpoint === 'explorer' ? 'rpc' : 'api']);
            const init = mergeRequestInit({
                body: method === 'POST' ? JSON.stringify(payload) : undefined,
                headers: {
                    'Accept-Encoding': 'gzip, deflate, br, zstd',
                    'Content-Type': 'application/json',
                },
                keepalive: true,
                method: method,
                signal: this.timeout ? AbortSignal.timeout(this.timeout) : undefined,
            }, this.fetchOptions, { signal });
            let request = new Request(url, init);
            if (this.onRequest) {
                const customRequest = await this.onRequest(request);
                if (customRequest instanceof Request)
                    request = customRequest;
            }
            let response = await fetch(request);
            if (this.onResponse) {
                const customResponse = await this.onResponse(response);
                if (customResponse instanceof Response)
                    response = customResponse;
            }
            if (!response.ok || !response.headers.get('Content-Type')?.includes('application/json')) {
                const body = await response.text().catch(() => undefined);
                throw new Error(body);
            }
            const body = await response.json();
            if (body?.type === 'error') {
                throw new Error(body?.message);
            }
            return body;
        }
        catch (error) {
            if (error instanceof Error)
                throw error;
            throw new Error(error);
        }
    }
}
