// @hotstuff-labs/ts-sdk/dist/transports/websocket.transport.js
import { ENDPOINTS_URLS } from '../utils';
import { TransportsTypes as TT } from '../types';
export class WebSocketTransport {
    constructor(options) {
        this.ws = null;
        this.reconnectAttempts = 0;
        this.maxReconnectAttempts = 5;
        this.reconnectDelay = 1000;
        this.messageQueue = [];
        this.messageIdCounter = 0;
        this.subscriptions = new Map();
        this.subscriptionCallbacks = new Map();
        this.connectionPromise = null;
        this.isTestnet = options?.isTestnet ?? false;
        this.timeout = options?.timeout ?? 10000;
        this.server = {
            mainnet: options?.server?.mainnet ?? ENDPOINTS_URLS.mainnet.ws,
            testnet: options?.server?.testnet ?? ENDPOINTS_URLS.testnet.ws,
        };
        this.keepAlive = options?.keepAlive ?? {
            interval: 30000,
            timeout: 10000,
        };
        const autoConnect = options?.autoConnect ?? true;
        if (autoConnect) {
            this.connectionPromise = this.connect().catch((error) => {
                console.error('Auto-connect failed:', error);
            });
        }
    }
    async ensureConnected() {
        if (this.ws?.readyState === WebSocket.OPEN) {
            return;
        }
        if (this.connectionPromise) {
            await this.connectionPromise;
            return;
        }
        this.connectionPromise = this.connect();
        await this.connectionPromise;
    }
    cleanup() {
        if (this.keepAliveInterval) {
            clearInterval(this.keepAliveInterval);
            this.keepAliveInterval = undefined;
        }
        this.messageQueue.forEach(({ reject }) => {
            reject(new Error('WebSocket disconnected'));
        });
        this.messageQueue = [];
    }
    startKeepAlive() {
        if (this.keepAlive?.interval) {
            this.keepAliveInterval = setInterval(() => {
                this.ping();
            }, this.keepAlive.interval);
        }
    }
    handleIncomingMessage(message) {
        if (message.id && (message.result !== undefined || message.error !== undefined)) {
            this.handleJSONRPCResponse(message);
            return;
        }
        if (message.method && message.params && !message.id) {
            this.handleJSONRPCNotification(message);
            return;
        }
    }
    handleJSONRPCResponse(response) {
        const queueItem = this.messageQueue.find((item) => item.id === response.id);
        if (queueItem) {
            this.messageQueue = this.messageQueue.filter((item) => item.id !== response.id);
            if (response.error) {
                queueItem.reject(new Error(`JSON-RPC Error ${response.error.code}: ${response.error.message}`));
            }
            else {
                queueItem.resolve(response.result);
            }
        }
        else {
        }
    }
    handleJSONRPCNotification(notification) {
        if ((notification.method === 'subscription' || notification.method === 'event') &&
            notification.params) {
            const { channel, data } = notification.params;
            let notificationSent = false;
            this.subscriptions.forEach((subscription, subscriptionId) => {
                if (subscription.channel === channel) {
                    const callback = this.subscriptionCallbacks.get(subscriptionId);
                    if (callback) {
                        callback({
                            channel,
                            data,
                            timestamp: Date.now(),
                        });
                        notificationSent = true;
                    }
                    else {
                        console.warn(`No callback found for subscription ${subscriptionId}`);
                    }
                }
            });
        }
        else {
            console.warn(`Unknown notification method: ${notification.method}`);
        }
    }
    async sendJSONRPCMessage(message) {
        await this.ensureConnected();
        return new Promise((resolve, reject) => {
            if (message.id === null || message.id === undefined) {
                message.id = (++this.messageIdCounter).toString();
            }
            if (message.id !== undefined) {
                this.messageQueue.push({
                    id: message.id,
                    resolve,
                    reject,
                });
                if (this.timeout) {
                    setTimeout(() => {
                        this.messageQueue = this.messageQueue.filter((qm) => qm.id !== message.id);
                        reject(new Error('Request timeout'));
                    }, this.timeout);
                }
            }
            const messageStr = JSON.stringify(message);
            this.ws.send(messageStr);
            if (message.id === undefined) {
                resolve({});
            }
        });
    }
    formatSubscriptionParams(channel, payload) {
        const payloadObj = payload;
        const subscription = {
            channel: channel,
            ...payloadObj,
        };
        // if (payloadObj?.instrumentId || payloadObj?.symbol) {
        //   subscription.symbol = payloadObj.instrumentId || payloadObj.symbol;
        // }
        // if (payloadObj?.user || payloadObj?.account) {
        //   subscription.user = payloadObj.user || payloadObj.account;
        // }
        return subscription;
    }
    async subscribeToChannels(params) {
        const message = {
            jsonrpc: '2.0',
            method: TT.WSMethod.SUBSCRIBE,
            params: params,
            id: (++this.messageIdCounter).toString(),
        };
        return this.sendJSONRPCMessage(message);
    }
    async unsubscribeFromChannels(channels) {
        const message = {
            jsonrpc: '2.0',
            method: TT.WSMethod.UNSUBSCRIBE,
            params: channels, // Send channels directly as array, not wrapped in object
            id: (++this.messageIdCounter).toString(),
        };
        return this.sendJSONRPCMessage(message);
    }
    isConnected() {
        return this.ws?.readyState === WebSocket.OPEN;
    }
    async connect() {
        return new Promise((resolve, reject) => {
            try {
                const url = this.isTestnet ? this.server.testnet : this.server.mainnet;
                this.ws = new WebSocket(url);
                this.ws.onopen = () => {
                    this.reconnectAttempts = 0;
                    this.startKeepAlive();
                    resolve();
                };
                this.ws.onmessage = (event) => {
                    try {
                        const message = JSON.parse(event.data);
                        this.handleIncomingMessage(message);
                    }
                    catch (error) {
                        console.error('Failed to parse WebSocket message:', error);
                    }
                };
                this.ws.onclose = (event) => {
                    this.cleanup();
                    if (this.reconnectAttempts < this.maxReconnectAttempts) {
                        setTimeout(() => {
                            this.reconnectAttempts++;
                            this.connect();
                        }, this.reconnectDelay * this.reconnectAttempts);
                    }
                };
                this.ws.onerror = (error) => {
                    console.error('WebSocket error:', error);
                    reject(error);
                };
            }
            catch (error) {
                reject(error);
            }
        });
    }
    async disconnect() {
        this.cleanup();
        if (this.ws) {
            this.ws.close();
            this.ws = null;
        }
    }
    async ping() {
        await this.ensureConnected();
        const message = {
            jsonrpc: '2.0',
            method: TT.WSMethod.PING,
            id: (++this.messageIdCounter).toString(),
        };
        return this.sendJSONRPCMessage(message);
    }
    async request(type, payload, signal) {
        throw new Error('Use specific WebSocket methods (subscribe, unsubscribe, ping) instead of generic request');
    }
    async subscribe(channel, payload, listener) {
        await this.ensureConnected();
        const subscriptionId = `${channel}_${Date.now()}`;
        const subscription = {
            id: subscriptionId,
            channel,
            symbol: payload?.instrumentId || payload?.symbol,
            params: payload,
            timestamp: Date.now(),
        };
        this.subscriptionCallbacks.set(subscriptionId, (data) => {
            const customEvent = new CustomEvent('subscription', {
                detail: data.data,
            });
            listener(customEvent);
        });
        try {
            const subscriptionParams = this.formatSubscriptionParams(channel, payload);
            const result = await this.subscribeToChannels(subscriptionParams);
            if (result.status === 'subscribed' && result.channels && result.channels.length > 0) {
                const serverChannel = result.channels[0];
                subscription.channel = serverChannel;
                this.subscriptions.set(subscriptionId, subscription);
                return { ...result, subscriptionId };
            }
            else if (result.status === 'failed' || result.status === 'error') {
                this.subscriptionCallbacks.delete(subscriptionId);
                const errorMsg = result.error || `Subscription ${result.status}`;
                throw new Error(`Server rejected subscription: ${errorMsg}`);
            }
            else if (!result.channels || result.channels.length === 0) {
                this.subscriptionCallbacks.delete(subscriptionId);
                throw new Error(`Server returned no channels for subscription`);
            }
            else {
                this.subscriptionCallbacks.delete(subscriptionId);
                throw new Error(`Unknown subscription status: ${result.status}`);
            }
        }
        catch (error) {
            this.subscriptions.delete(subscriptionId);
            this.subscriptionCallbacks.delete(subscriptionId);
            console.error(`Subscription failed for ${channel}:`, error);
            throw error;
        }
    }
    async unsubscribe(subscriptionId) {
        const subscription = this.subscriptions.get(subscriptionId);
        if (!subscription) {
            throw new Error(`Subscription ${subscriptionId} not found`);
        }
        try {
            if (this.isConnected()) {
                await this.unsubscribeFromChannels([subscription.channel]);
            }
            this.subscriptions.delete(subscriptionId);
            this.subscriptionCallbacks.delete(subscriptionId);
        }
        catch (error) {
            console.error(`Failed to unsubscribe from ${subscription.channel}:`, error);
            this.subscriptions.delete(subscriptionId);
            this.subscriptionCallbacks.delete(subscriptionId);
            throw error;
        }
    }
    getSubscriptions() {
        return Array.from(this.subscriptions.values());
    }
}
